
# biomasslmb

<!-- badges: start -->
<!-- badges: end -->

The goal of `biomasslmb` is to provide functions to pipe together proteomics data analysis pipelines for data generated by the [Biological Mass Spectrometry and Proteomics Facility](https://www2.mrc-lmb.cam.ac.uk/research/scientific-facilities-and-support-services/mass-spectrometry-home/) at the [MRC Laboratory of Molecular Biology](https://www2.mrc-lmb.cam.ac.uk/).

`biomasslmb` relies upon the gold-standard Bioconductor package for quantitative proteomics [`QFeatures`](https://www.bioconductor.org/packages/release/bioc/html/QFeatures.html). 

This package also provides some more esoteric and/or opinionated functionality for proteomics data analysis. Much of the initial code is a re-working of [`camprotR`](https://github.com/CambridgeCentreForProteomics/camprotR) which we co-developed by Charlotte Dawson (https://github.com/csdaw) and myself (https://github.com/tomsmithCGAT) and was designed to work with [`MSnbase`](https://bioconductor.org/packages/release/bioc/html/MSnbase.html). Hence I've reworked the code here as a separate package to simply issues with backward compatibility. Thank you to Charlotte for any of the code lifted from `camprotR` for which I had no part in writing!

## Installation

You can install the development version of biomasslmb like so:

``` r
remotes::install_github("lmb-mass-spec-compbio/biomasslmb", dependencies = TRUE)
```

## Example

Below is a routine work flow for generating
accurate protein-level abundances from PSM level quantification for TMT proteomics.

``` r


library(QFeatures)
library(biomasslmb)

# Read in PSM-level quantification from TMT experiment (using QFeatures function)
tmt_qf <- readQFeatures(assayData = psm_tmt_total,
                        quantCols = 36:45, 
                        name = "psms_raw")

crap_fasta_inf <- system.file(
  "extdata", "cRAP_20190401.fasta.gz",
  package = "biomasslmb"
)

crap_accessions <- get_crap_fasta_accessions(crap_fasta_inf)

# Perform routine raw data filtering.
# - Remove PSMs from contaminant proteins
# - Remove PSMs where protein ID is empty or not unique
tmt_qf[['psms_raw']] <- filter_features(tmt_qf[['psms_raw']], crap_proteins=crap_accessions)

# Add more accurate average S:N ratio value
tmt_qf[['psms_raw']] <- update_average_sn(tmt_qf[['psms_raw']])

# Filter PSMs to remove low S:N and/or high interference
tmt_qf[['psms_filtered']] <- filter_TMT_PSMs(tmt_qf[['psms_raw']], inter_thresh=50, sn_thresh=5)

# Plot the missing values in relation to S:N
plot_missing_SN(tmt_qf[['psms_filtered']])
plot_missing_SN_per_sample(tmt_qf[['psms_filtered']])

# Add information about where the peptide sequence is within the protein sequence
# For this, we need a fasta file with the protein sequences
protein_ids <- rowData(tmt_qf[['psms_filtered']])$Master.Protein.Accessions
temp_fasta_file <- tempfile()
make_fasta(protein_ids, temp_fasta_file)

tmt_qf[['psms_filtered']] <- add_peptide_positions(tmt_qf[['psms_filtered']], proteome_fasta=temp_fasta_file)


# Remove PSMs for proteins with fewer than 2 PSMs
min_psms <- 2
tmt_qf[['psms_filtered']] <- filter_features_per_protein(tmt_qf[['psms_filtered']], min_features = min_psms)

# Plot the PSM-level quantification distributions per sample
plot_quant(tmt_qf[['psms_filtered']], log2transform=TRUE, method='density')

# Aggregate to protein-level abundances (using QFeatures function)
tmt_qf <- aggregateFeatures(tmt_qf, 
                            i = "psms_filtered", 
                            fcol = "Master.Protein.Accessions",
                            name = "Protein",
                            fun = base::colSums,
                            na.rm = TRUE)

# Replace protein-level quantification values with NA where they derive from fewer
# than 2 finite PSM-level quantification values
protein_mask <- get_protein_no_quant_mask(tmt_qf[['psms_filtered']], min_features = min_psms, plot=TRUE)
tmt_qf[['Protein']] <- mask_protein_level_quant(tmt_qf[['Protein']], protein_mask)

# Plot the Protein-level quantification distributions per sample
plot_quant(tmt_qf[['Protein']], log2transform=TRUE, method='density')

```

