---
title: "LFQ-DDA QC Peptide-level quantification and summarisation to protein-level abundance"
author: "Tom Smith"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
bibliography: biomasslmb.json
vignette: >
  %\VignetteIndexEntry{LFQ-DDA QC Peptide-level quantification and summarisation to protein-level abundance}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```


Label-Free Quantification (LFQ) is the simplest form of quantitative proteomics, in which different samples are quantified in separate MS runs. Quantification is either performed by Data-Dependent Aquisition (DDA), where the Mass Spectrometer triggers fragmentation of ions within a given m/z range with the aim being to focus attention of individual peptides separately, or Data-Independent Aquisition (DIA), where a much wider m/z range is used and a mix of peptides are co-fragmented and quantified simultaneously by deconvoluting the resultant complex spectra. Here, we will focus on just LFQ-DDA.

Since each sample is run separately, different peptides will be quantified in each sample and the peptide intensities may not be directly comparable between samples. The common solution to the higher burden of missing values is to use the ‘match-between-runs’ (Cox et al. 2014[UPDATE REF!!!]), or the functionally equivalent ‘Minora’ algorithm employed by Proteome Discoverer (PD). These algorithms use the observed retention times of MS1 ions which were successfully spectrum matched in one sample to identify the likely peptide sequence of MS1 ions that could not be spectrum matched in another sample. However, even with these algorithms, DDA LFQ will still typically have many more missing values than labelled LFQ, e.g TMT.

Despite the pitfalls of LFQ, the data analysis is still relatively straightforward, though there are steps that need some careful consideration and quality control assessment.

# Load required packages

To clarify which functionality is provided by which package, we will use `package::function`. For your own code, there is no need to specify the package unless you want to maintain this clarity.

```{r, message=FALSE}
library(QFeatures)
library(biomasslmb)
library(ggplot2)
library(tidyr)
library(dplyr)
```

# Defining the contaminant proteins

We need to remove contaminant proteins. These were defined here using the cRAP database. Below, we parse the contaminants fasta to extract the IDs for the proteins in both 'cRAP' format and Uniprot IDs.

```{r}

crap_fasta_inf <- system.file(
  "extdata", "cRAP_20190401.fasta.gz", 
  package = "biomasslmb"
)

# Extract the protein IDs associated with each cRAP protein
crap_accessions <- biomasslmb::get_crap_fasta_accessions(crap_fasta_inf)

print(head(crap_accessions))
```

# Read in input data

We start by reading in quantitative proteomics data into a `QFeatures` object, which is the standard Bioconductor object for holding quantitative proteomics data. See [here](https://www.bioconductor.org/packages/release/bioc/html/QFeatures.html) for documentation about the `QFeatures` object.In this case, we are not adding any experimental details to the `QFeatures` object, so the `ColData` will be empty.

`??` is a data set available from the `biomasslmb` package 

`lfq_dda_pd_PeptideGroups.txt` is a file available from the `biomasslmb` package containing the Peptide-level output from Proteome Discoverer (PD) for an experiment with 6 samples. It is a truncated file containing the precursors peptides quantification for just 500 proteins. 

```{r}
pep_inf <- system.file(
  "extdata", "lfq_dda_pd_PeptideGroups.txt", 
  package = "biomasslmb"
)

```

Inspecting the column names for the abundances, we see the names are not particularly useful. The `Abundance.FXX.na.Sample.` part is irrelevant and the replicate number is not included
```{r}
infdf <- read.delim(pep_inf)

abundance_cols_ix <-18:23

colnames(infdf)[abundance_cols_ix]
```

Below, we update the column names for the abundance columns
```{r}
# Clean up sample names
colnames(infdf)[abundance_cols_ix] <- paste(gsub('Abundance.F\\d+.na.Sample.', '',
                                        colnames(infdf)[abundance_cols_ix]), rep(1:3, times=2), sep='_')

colnames(infdf)[abundance_cols_ix]
```

Before we read in the data, we also want to define the experimental conditions for each sample. We can then provide the 

```{r}
exp_design <- data.frame(quantCols=colnames(infdf)[abundance_cols_ix]) %>%
  separate(quantCols, sep='_', into=c('condition', 'replicate'), remove = FALSE) 

knitr::kable(exp_design)
```

Now we can read the data into a `QFeatures` object
```{r}
# Read in PSM-level quantification from TMT experiment (using QFeatures function)
lfq_qf <- readQFeatures(assayData = infdf,
                        quantCols = abundance_cols_ix, 
                        colData = exp_design,
                        name = "peptides_raw")

```

Adding the `colData` to the peptide-level data too.

```{r}
colData(lfq_qf[['peptides_raw']]) <- colData(lfq_qf)


```

Extract the gene name and long-form protein name from the descriptions.
```{r, eval=FALSE}

tmt_qf[['peptides_raw']] <- add_gene_long_protein_name_pd(tmt_qf[['peptides_raw']])
```


It's helpful to assess the retention time (RT) for the peptides and the
relationship with the delta mass for the precursor ion. Everything looks
OK here.

```{r, warning=FALSE, fig.show='hold', fig.show="hold", out.width="50%"}
plot_rt_dist(lfq_qf[['peptides_raw']])
plot_rt_vs_delta(lfq_qf[['peptides_raw']])
```


# Filter peptides

We first perform routine filtering to remove PSMs that:

-   Could originate from contaminants. See `?filter_features` for
    further details, including the removal of 'associated cRAP'.
-   Don't have a unique master protein
-   Don't have any quantification values

```{r}
# Perform routine raw data filtering.
# - Remove PSMs from contaminant proteins
# - Remove PSMs where protein ID is empty or not unique

lfq_qf[['peptides_filtered']] <- filter_features_pd_dda(lfq_qf[['peptides_raw']],
                                                        protein_col='Master.Protein.Accessions', # UPDATE WHEN RE-EXPORTED
                                                 contaminant_proteins=crap_accessions,
                                                 filter_crap=TRUE,
                                                 filter_associated_crap=TRUE,
                                                 unique_master=FALSE,
                                                 remove_no_quant = TRUE)

```


We will also remove peptides that are not rank 1 according to the search
engine

```{r}

lfq_qf <- lfq_qf %>%
filterFeatures(~ Search.Engine.Rank.by.Search.Engine.Sequest.HT == 1,
               i = "peptides_filtered") 

biomasslmb:::message_parse(rowData(lfq_qf[['peptides_filtered']]),
                           'Master.Protein.Accessions',
                           "Removing peptides that are not rank 1")
```

For the normalisation and summarisation, we need our quantification
values to be on a log-scale, so we log-transform below

```{r}
lfq_qf[['peptides_filtered']] <- QFeatures::logTransform(
  lfq_qf[['peptides_filtered']], base=2)
```

Next, we plot the peptide intensities.

```{r, warning=FALSE, fig.show='hold', fig.show="hold", out.width="50%"}

# Plot the peptide-level quantification distributions per sample
plot_quant(lfq_qf[['peptides_filtered']], log2transform=FALSE, method='density') +
  scale_colour_manual(values=rep(get_cat_palette(3), each=3)) +
  aes(linetype=sample) +
  scale_linetype_manual(values=rep(1:3, times=3)) +
  xlab('Peptide abundance (log2)')
```




## Normalise

Since the same amount of sample was injected for each sample, we want to
normalise to remove differences in the overall disribution of
abundances. We will thus perform 'diff.median' normalisation with
`QFeatures::normalize`.

```{r}
lfq_qf[['peptides_norm']] <- QFeatures::normalize(lfq_qf[['peptides_filtered']],
                                                  method='diff.median')
```

```{r, warning=FALSE, fig.show='hold', fig.show="hold", out.width="50%"}

# Plot the peptide-level quantification distributions per sample
plot_quant(lfq_qf[['peptides_norm']], log2transform=FALSE, method='density') +
  scale_colour_manual(values=rep(get_cat_palette(3), each=3)) +
  aes(linetype=sample) +
  scale_linetype_manual(values=rep(1:3, times=3)) +
  xlab('Peptide abundance (log2)')
```


## Summarising to protein-level abundances

For the summarisation to protein, we will use `robustSummary`. We do not need to remove
peptides with any missing values. However, we still don't want to retain
peptides with too many missing values, since these will not be very
informative in estimating the protein-level quantification.

Below, we inspect the most common patterns for the missing values.

The missingness is most frequently consigned to all samples from one condition and also commonly 2/3 replicates from one condition. It looks like the missingness is far from missing at random (MAR) and is likely missing due to falling below detection limits.

```{r, fig.height=6, fig.width=8}
plot_missing_upset(lfq_qf, i='peptides_norm')
```

Since DDA LFQ typically has too many missing values to exclude peptides with missing values, the `MsCoreUtils::robustSummary` method should preferred for summarisation to protein-level abundance. With `robustSummary`, we do not need to remove all PSMs with missing values since the summarisation algorithm deals with them appropriately  [@http://zotero.org/users/5634351/items/FZN3QTTZ].. However, we still don't want to retain PSMs with too many missing values, since these will not be very informative in estimating the protein-level quantification. Here, we will retain PSMs with at most 4/6 missing values 

```{r}

lfq_qf[['peptides_filtered_missing']] <- QFeatures::filterNA(
  lfq_qf[['peptides_norm']], 4/6)

biomasslmb:::message_parse(rowData(lfq_qf[['peptides_filtered_missing']]),
                           'Master.Protein.Accessions',
                           "Removing peptides with more than 4/6 missing values")
```

Next, we remove peptides for proteins with fewer than 2 peptides.

```{r}
min_peps <- 2
lfq_qf[['peptides_for_summarisation']] <- filter_features_per_protein(
  lfq_qf[['peptides_filtered_missing']], min_features = min_peps)

biomasslmb:::message_parse(rowData(lfq_qf[['peptides_for_summarisation']]),
                           'Master.Protein.Accessions',
                           "Removing 'one-hit' wonders")
```

Now we can summarise with `robustSummary`

```{r}
set.seed(42)

# Aggregate to protein-level abundances (using QFeatures function)
lfq_qf <- aggregateFeatures(lfq_qf, 
                            i = "peptides_for_summarisation",
                            #i = "peptides_filtered_missing", 
                            fcol = "Master.Protein.Accessions",
                            name = "protein",
                            fun = MsCoreUtils::robustSummary,
                            maxit=10000)

```

Prior to summaristaion, we removed PSMs from proteins with fewer than 2
peptides. However, since we left in peptides with missing values, it's
possible for some protein-level abundances to be derived from just a
single peptide. We can use the `get_protein_no_quant_mask` from
`biomasslmb` to identify where the protein abundances will be derived
from fewer than `n` features (peptides). We can then give this mask to
`mask_protein_level_quant` to replace these quantification values with
NA.

```{r, warning=FALSE, fig.show='hold', out.width="50%", fig.width=4, fig.height=4}
# plot = TRUE means we will also get a plot of the number of proteins quantified in each sample
protein_retain_mask <- biomasslmb::get_protein_no_quant_mask(
  lfq_qf[['peptides_for_summarisation']], min_features=min_peps, plot=TRUE) 

lfq_qf[['protein']] <- biomasslmb::mask_protein_level_quant(
  lfq_qf[['protein']], protein_retain_mask)

```

## Re-inspecting missing values at protein-level

Now that we have protein-level abundances, we would like to re-inspect the missing values. Overall, we have `r round(nNA(lfq_qf[['protein']])$nNA$pNA*100, 3)` % missing values, with at most `r round(max(nNA(dia_qf[['protein']])$nNAcols$pNA)*100, 3)` % missing values in any given sample. The most common missingness patterns are for the protein to be missing from 1,2 or 3 replicates from a single condition, suggesting the quantification is missing due to detection limits.

```{r, fig.height=6, fig.width=8}

plot_missing_upset(lfq_qf, i='protein' )
```


## Inspecting the number of peptides and proteins through the processing steps

Now that we have performed all the filtering steps and summarisation to protein-level abundances, it's helpful to visualise how many peptides/proteins were retained at each level of the processing. We can use the `biomasslmb::get_samples_present` and `biomasslmb::plot_samples_present` functions for this. First though, we need to decide which 'experiments' we want to plot and define a named character vector since the QFeatures names are not sufficiently clear by themselves

Below, we inspect the experiment names.

```{r}
names(lfq_qf)
```

In this case, we don't want to plot the `peptides_norm`, since that's just the normalised quantification present in `peptides_filtered` and no peptides were removed in that step.

### Samples per peptide

We'll start by inspecting the number of peptides in each experiment. We therefore define a named character vector all the peptides-level experiments, excluding `psms_filtered_norm`. We set the row variables to be `Annotated.Sequence` and `Modifications` so that we count the number of unique peptides.

```{r, fig.height=7, , fig.width=7, fig.fullwidth=TRUE, fig.cap="Samples quantified for each PSM at each level of processing", out.width='75%'}


rename_cols <- c('All peptides' = 'peptides_raw' ,
                 'Quantified, contaminants removed' = 'peptides_filtered',
                 'No missing values' = 'peptides_filtered_missing',
                 '>1 PSM per protein' = 'peptides_for_summarisation')

rowvars <- c('Annotated.Sequence', 'Modifications')

samples_present <- get_samples_present(lfq_qf[,,unname(rename_cols)], rowvars, rename_cols)
plot_samples_present(samples_present, rowvars, breaks=seq(2,10,2)) + ylab('PSM')



```

### Samples per Protein

Next, we'll use the same functions to inspect the number of proteins at each level of processing. We need to supply an updated named character vector to include the `protein` experiment and set the row variables to be just the `Master.Protein.Accesions` column.

```{r, fig.height=7, , fig.width=7, fig.fullwidth=TRUE, fig.cap="Samples quantified for each protein at each level of processing", out.width='75%'}

rename_cols_prot <- c(rename_cols, 'Protein'='protein')

rowvars_prot <- c('Master.Protein.Accessions')

samples_present <- get_samples_present(lfq_qf, rowvars_prot, rename_cols_prot)
plot_samples_present(samples_present, rowvars_prot, breaks=seq(2,10,2))
```

From these two plots, we can see that the filtering to ensure that all proteins have \>1 peptides removed only a few peptides, but more proteins. Whether this is appropriate will depend on your data in hand.


```{r, include = FALSE}
# Save file to package as data so it can be read back in in other vignettes
usethis::use_data(lfq_qf, overwrite=TRUE)
```

```{r}
sessionInfo()
```
