<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Package index â€¢ biomasslmb</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Package index"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">biomasslmb</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/LFQ_DDA_Peptide_QC_Summarisation.html">LFQ-DDA QC Peptide-level quantification and summarisation to protein-level abundance</a></li>
    <li><a class="dropdown-item" href="../articles/LFQ_DIA_Peptide_QC.html">LFQ-DIA QC Peptide-level quantification and summarisation to protein-level abundance</a></li>
    <li><a class="dropdown-item" href="../articles/TMT_PSM_QC_Summarisation.html">TMT QC PSM-level quantification and summarisation to protein-level abundance</a></li>
    <li><a class="dropdown-item" href="../articles/TMT_summarisation_methods.html">TMT - comparing protein summarisation approaches</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Package index</h1>
    </div>

    <div class="section level2">
      <h2 id="all-functions">All functions<a class="anchor" aria-label="anchor" href="#all-functions"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="add_gene_long_protein_name_pd.html">add_gene_long_protein_name_pd()</a></code>

        </dt>
        <dd>Extract the gene name and long-form protein name from the master protein descriptions column</dd>
      </dl><dl><dt>

          <code><a href="add_independent_filtering_padj.html">add_independent_filtering_padj()</a></code>

        </dt>
        <dd>Perform independent filtering to threshold the over-representation testing on the number of features in each category to limit the multiple testing burden</dd>
      </dl><dl><dt>

          <code><a href="add_peptide_positions.html">add_peptide_positions()</a></code>

        </dt>
        <dd>Add a column describing the position of the peptide sequence with respect to the protein</dd>
      </dl><dl><dt>

          <code><a href="add_peptide_positions_from_cleavage.html">add_peptide_positions_from_cleavage()</a></code>

        </dt>
        <dd>Add peptide positions with respect to the protein</dd>
      </dl><dl><dt>

          <code><a href="add_ptm_pos_rowdata_mq.html">add_ptm_pos_rowdata_mq()</a></code>

        </dt>
        <dd>Add rowData columns with positions of PTMs with respect to peptide sequence</dd>
      </dl><dl><dt>

          <code><a href="add_ptm_positions.html">add_ptm_positions()</a></code>

        </dt>
        <dd>Add rowData columns with details of PTMs positions</dd>
      </dl><dl><dt>

          <code><a href="add_tm_info.html">add_tm_info()</a></code>

        </dt>
        <dd>Add transmembrane domain details</dd>
      </dl><dl><dt>

          <code><a href="add_topology_info.html">add_topology_info()</a></code>

        </dt>
        <dd>Add topology details</dd>
      </dl><dl><dt>

          <code><a href="append_fasta.html">append_fasta()</a></code>

        </dt>
        <dd>Append sequences to end of a FASTA</dd>
      </dl><dl><dt>

          <code><a href="center_normalise_to_ref.html">center_normalise_to_ref()</a></code>

        </dt>
        <dd>Center-median normalise the expression matrix in an MSnSet using medians from a reference dataset</dd>
      </dl><dl><dt>

          <code><a href="check_uniprot_release.html">check_uniprot_release()</a></code>

        </dt>
        <dd>Check the current UniProt release</dd>
      </dl><dl><dt>

          <code><a href="create_long_form_imputed_data.html">create_long_form_imputed_data()</a></code>

        </dt>
        <dd>Create long format data with column defining if quantification value is imputed</dd>
      </dl><dl><dt>

          <code><a href="dia_qf.html">dia_qf</a></code>

        </dt>
        <dd>DIA data</dd>
      </dl><dl><dt>

          <code><a href="estimate_overrep.html">estimate_overrep()</a></code>

        </dt>
        <dd>Estimate effect size of over-representation</dd>
      </dl><dl><dt>

          <code><a href="filter_TMT_PSMs.html">filter_TMT_PSMs()</a></code>

        </dt>
        <dd>Filter a PSM-level summarizedExperiment to remove low quality PSMs</dd>
      </dl><dl><dt>

          <code><a href="filter_by_protein_fdr.html">filter_by_protein_fdr()</a></code>

        </dt>
        <dd>Filter to remove peptides from proteins failing the protein FDR threshold</dd>
      </dl><dl><dt>

          <code><a href="filter_features_diann.html">filter_features_diann()</a></code>

        </dt>
        <dd>Filter DIA-NN output</dd>
      </dl><dl><dt>

          <code><a href="filter_features_pd_dda.html">filter_features_pd_dda()</a></code>

        </dt>
        <dd>Filter Proteome Discoverer DDA output</dd>
      </dl><dl><dt>

          <code><a href="filter_features_per_protein.html">filter_features_per_protein()</a></code>

        </dt>
        <dd>Remove features which are assigned to a protein with too few supporting features in total</dd>
      </dl><dl><dt>

          <code><a href="filter_features_sn.html">filter_features_sn()</a></code>

        </dt>
        <dd>Filter Spectronaut output</dd>
      </dl><dl><dt>

          <code><a href="filter_maxquant_ptm.html">filter_maxquant_ptm()</a></code>

        </dt>
        <dd>Make a FASTA using UniProt accessions</dd>
      </dl><dl><dt>

          <code><a href="get_ancestor_go.html">get_ancestor_go()</a></code>

        </dt>
        <dd>Get all ancestor GO terms</dd>
      </dl><dl><dt>

          <code><a href="get_cat_palette.html">get_cat_palette()</a></code>

        </dt>
        <dd>Generate a colour-blind friendly palette for categorical colour encoding</dd>
      </dl><dl><dt>

          <code><a href="get_crap_fasta_accessions.html">get_crap_fasta_accessions()</a></code>

        </dt>
        <dd>Extract the cRAP protein accessions from a fasta file</dd>
      </dl><dl><dt>

          <code><a href="get_enriched_go.html">get_enriched_go()</a></code>

        </dt>
        <dd>GO term enrichment using goseq</dd>
      </dl><dl><dt>

          <code><a href="get_go_terms.html">get_go_terms()</a></code>

        </dt>
        <dd>Obtain GO term annotations for proteins</dd>
      </dl><dl><dt>

          <code><a href="get_medians.html">get_medians()</a></code>

        </dt>
        <dd>Extract the assay column medians from an MSnSet</dd>
      </dl><dl><dt>

          <code><a href="get_n_feature_per_prot.html">get_n_feature_per_prot()</a></code>

        </dt>
        <dd>Identify how many features (PSMs/Peptides) are quantified for each protein</dd>
      </dl><dl><dt>

          <code><a href="get_protein_no_quant_mask.html">get_protein_no_quant_mask()</a></code>

        </dt>
        <dd>Identify proteins which have too few features to quantify protein abundance in each sample</dd>
      </dl><dl><dt>

          <code><a href="get_protein_tm_topology.html">get_protein_tm_topology()</a></code>

        </dt>
        <dd>Obtain protein transmembrane-domains and topology information</dd>
      </dl><dl><dt>

          <code><a href="get_samples_present.html">get_samples_present()</a></code>

        </dt>
        <dd>Extract the number of samples each feature was detected in for each experiment in a Qfeatures object</dd>
      </dl><dl><dt>

          <code><a href="make_fasta.html">make_fasta()</a></code>

        </dt>
        <dd>Make a FASTA using UniProt accessions</dd>
      </dl><dl><dt>

          <code><a href="mask_protein_level_quant.html">mask_protein_level_quant()</a></code>

        </dt>
        <dd>Replace protein-level quantifications with NA if they derive from too few lower feature level quantifications</dd>
      </dl><dl><dt>

          <code><a href="plot_cor_samples.html">plot_cor_samples()</a></code>

        </dt>
        <dd>Plot the correlation between sample</dd>
      </dl><dl><dt>

          <code><a href="plot_go.html">plot_go()</a></code>

        </dt>
        <dd>Plot selected GO terms</dd>
      </dl><dl><dt>

          <code><a href="plot_go_terms_upset.html">plot_go_terms_upset()</a></code>

        </dt>
        <dd>Plot selected GO terms</dd>
      </dl><dl><dt>

          <code><a href="plot_missing_SN.html">plot_missing_SN()</a></code>

        </dt>
        <dd>Plot the missing values vs signal:noise</dd>
      </dl><dl><dt>

          <code><a href="plot_missing_SN_per_sample.html">plot_missing_SN_per_sample()</a></code>

        </dt>
        <dd>Plot the missing values vs signal:noise for each sample</dd>
      </dl><dl><dt>

          <code><a href="plot_missing_upset.html">plot_missing_upset()</a></code>

        </dt>
        <dd>Plot the most common missing value patterns</dd>
      </dl><dl><dt>

          <code><a href="plot_pca.html">plot_pca()</a></code>

        </dt>
        <dd>Create a Principle Component plot from the feature quantification</dd>
      </dl><dl><dt>

          <code><a href="plot_protein_assays.html">plot_protein_assays()</a></code>

        </dt>
        <dd>Plot quantification at multiple levels</dd>
      </dl><dl><dt>

          <code><a href="plot_quant.html">plot_quant()</a></code>

        </dt>
        <dd>Plot distributions for feature intensities per sample.</dd>
      </dl><dl><dt>

          <code><a href="plot_rt_dist.html">plot_rt_dist()</a></code>

        </dt>
        <dd>Plot the retention time distribution</dd>
      </dl><dl><dt>

          <code><a href="plot_rt_vs_delta.html">plot_rt_vs_delta()</a></code>

        </dt>
        <dd>Plot the retention time vs delta precursor mass</dd>
      </dl><dl><dt>

          <code><a href="plot_samples_present.html">plot_samples_present()</a></code>

        </dt>
        <dd>Plots the number of samples each feature was detected in for each experiment in a Qfeatures object</dd>
      </dl><dl><dt>

          <code><a href="plot_volcano.html">plot_volcano()</a></code>

        </dt>
        <dd>Volcano plot for differential abundance testing</dd>
      </dl><dl><dt>

          <code><a href="psm_tmt_total.html">psm_tmt_total</a></code>

        </dt>
        <dd>PSM-level PD output for total proteome TMT10-plex data</dd>
      </dl><dl><dt>

          <code><a href="query_protein_tm_topology.html">query_protein_tm_topology()</a></code>

        </dt>
        <dd>Query UniProt to determine protein transmembrane-domains and topology</dd>
      </dl><dl><dt>

          <code><a href="readDIANNFilterQJoin.html">readDIANNFilterQJoin()</a></code>

        </dt>
        <dd>Read in data from DIA-NN</dd>
      </dl><dl><dt>

          <code><a href="remove_dots.html">remove_dots()</a></code>

        </dt>
        <dd>Remove duplicated full stops</dd>
      </dl><dl><dt>

          <code><a href="remove_redundant_go.html">remove_redundant_go()</a></code>

        </dt>
        <dd>Remove redundant GO terms</dd>
      </dl><dl><dt>

          <code><a href="remove_redundant_psm_quant.html">remove_redundant_psm_quant()</a></code>

        </dt>
        <dd>Remove redundant Peptide Spectrum Matches</dd>
      </dl><dl><dt>

          <code><a href="remove_x.html">remove_x()</a></code>

        </dt>
        <dd>Remove leading X</dd>
      </dl><dl><dt>

          <code><a href="restrict_imputation.html">restrict_imputation()</a></code>

        </dt>
        <dd>Restrict imputed values to specific conditions</dd>
      </dl><dl><dt>

          <code><a href="sub_crap.html">sub_crap()</a></code>

        </dt>
        <dd>Insert cRAP numbers into a character vector</dd>
      </dl><dl><dt>

          <code><a href="theme_biomasslmb.html">theme_biomasslmb()</a></code>

        </dt>
        <dd>A ggplot2 theme for the package</dd>
      </dl><dl><dt>

          <code><a href="tmt_qf.html">tmt_qf</a></code>

        </dt>
        <dd>TMT data</dd>
      </dl><dl><dt>

          <code><a href="update_average_sn.html">update_average_sn()</a></code>

        </dt>
        <dd>Adds new feature describing the average reporter Signal/Noise ratio.</dd>
      </dl><dl><dt>

          <code><a href="update_peptide_assignments.html">update_peptide_assignments()</a></code>

        </dt>
        <dd>Update PSM level protein/master protein assignments using peptide-level assignments</dd>
      </dl></div>
  </main></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Tom Smith.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

