<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>LFQ-DIA QC Peptide-level quantification and summarisation to protein-level abundance • biomasslmb</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="LFQ-DIA QC Peptide-level quantification and summarisation to protein-level abundance">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">biomasslmb</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/LFQ_DDA_Peptide_QC_Summarisation.html">LFQ-DDA QC Peptide-level quantification and summarisation to protein-level abundance</a></li>
    <li><a class="dropdown-item" href="../articles/LFQ_DIA_Peptide_QC.html">LFQ-DIA QC Peptide-level quantification and summarisation to protein-level abundance</a></li>
    <li><a class="dropdown-item" href="../articles/TMT_PSM_QC_Summarisation.html">TMT QC PSM-level quantification and summarisation to protein-level abundance</a></li>
    <li><a class="dropdown-item" href="../articles/TMT_summarisation_methods.html">TMT - comparing protein summarisation approaches</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>LFQ-DIA QC Peptide-level quantification and summarisation to protein-level abundance</h1>
                        <h4 data-toc-skip class="author">Tom Smith</h4>
            
            <h4 data-toc-skip class="date">2025-04-02</h4>
      

      <div class="d-none name"><code>LFQ_DIA_Peptide_QC.Rmd</code></div>
    </div>

    
    
<p>THIS VIGNETTE IS A WORK IN PROGRESS!</p>
<p>Label-Free Quantification (LFQ) is the simplest form of quantitative
proteomics, in which different samples are quantified in separate MS
runs. Quantification is either performed by Data-Dependent Aquisition
(DDA), where the Mass Spectrometer triggers fragmentation of ions within
a given m/z range with the aim being to focus attention of individual
peptides separately, or Data-Independent Aquisition (DIA), where a much
wider m/z range is used and a mix of peptides are co-fragmented and
quantified simultaneously by deconvoluting the resultant complex
spectra. Here, we will focus on just LFQ-DIA.</p>
<p>MORE DETAILS ON DIA REQUIRED HERE!</p>
<div class="section level2">
<h2 id="load-required-packages">Load required packages<a class="anchor" aria-label="anchor" href="#load-required-packages"></a>
</h2>
<p>To clarify which functionality is provided by which package, we will
use <code>package::function</code>. For your own code, there is no need
to specify the package unless you want to maintain this clarity.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/RforMassSpectrometry/QFeatures" class="external-link">QFeatures</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://lmb-mass-spec-compbio.github.io/biomasslmb/">biomasslmb</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org" class="external-link">tidyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="defining-the-contaminant-proteins">Defining the contaminant proteins<a class="anchor" aria-label="anchor" href="#defining-the-contaminant-proteins"></a>
</h2>
<p>We need to remove contaminant proteins. These were defined here using
the cRAP database. Below, we parse the contaminants fasta to extract the
IDs for the proteins in both ‘cRAP’ format and Uniprot IDs.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">crap_fasta_inf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span></span>
<span>  <span class="st">"extdata"</span>, <span class="st">"cRAP_20190401.fasta.gz"</span>, </span>
<span>  package <span class="op">=</span> <span class="st">"biomasslmb"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Extract the protein IDs associated with each cRAP protein</span></span>
<span><span class="va">crap_accessions</span> <span class="op">&lt;-</span> <span class="fu">biomasslmb</span><span class="fu">::</span><span class="fu"><a href="../reference/get_crap_fasta_accessions.html">get_crap_fasta_accessions</a></span><span class="op">(</span><span class="va">crap_fasta_inf</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">crap_accessions</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "cRAP001" "P00330"  "cRAP002" "P02768"  "cRAP003" "P00883"</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="read-in-input-data">Read in input data<a class="anchor" aria-label="anchor" href="#read-in-input-data"></a>
</h2>
<p>We start by reading in quantitative proteomics data into a
<code>QFeatures</code> object, which is the standard Bioconductor object
for holding quantitative proteomics data. See <a href="https://www.bioconductor.org/packages/release/bioc/html/QFeatures.html" class="external-link">here</a>
for documentation about the <code>QFeatures</code> object.</p>
<p>Here, we will use the <code>biomasslmb::readDIANNFilter</code>
function, which also allows us to control the FDR thresholds for the
precursor and protein (run-specific and global). This function reads in
the <code>report.tsv</code> file from DIA-NN. If you are happy with the
filtering settings in DIA-NN, you can directly read the
<code>report.pr_matrix.tsv</code> into a <code>QFeatures</code> object
using the <code><a href="https://rdrr.io/pkg/QFeatures/man/readQFeatures.html" class="external-link">QFeatures::readQFeatures</a></code> function.</p>
<p><code>diann_report.tsv</code> is a file available from the
<code>biomasslmb</code> package containing the output from DIA-NN for an
experiment with 6 samples. It is a truncated file containing the
precursors (peptides) for just 500 proteins.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">diann_report_inf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span></span>
<span>  <span class="st">"extdata"</span>, <span class="st">"diann_report.tsv"</span>, </span>
<span>  package <span class="op">=</span> <span class="st">"biomasslmb"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">dia_qf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readDIANNFilterQJoin.html">readDIANNFilterQJoin</a></span><span class="op">(</span><span class="va">diann_report_inf</span>,</span>
<span>                               global_protein_q<span class="op">=</span><span class="fl">0.01</span>,</span>
<span>                               run_protein_q<span class="op">=</span><span class="fl">0.01</span>,</span>
<span>                               run_precursor_q<span class="op">=</span><span class="fl">0.01</span><span class="op">)</span></span>
<span><span class="co">#&gt; Checking arguments.</span></span>
<span><span class="co">#&gt; Loading data as a 'SummarizedExperiment' object.</span></span>
<span><span class="co">#&gt; Splitting data in runs.</span></span>
<span><span class="co">#&gt; Formatting sample annotations (colData).</span></span>
<span><span class="co">#&gt; Formatting data as a 'QFeatures' object.</span></span>
<span><span class="co">#&gt; 'Lib.PG.Q.Value' found in 6 out of 6 assay(s)</span></span>
<span><span class="co">#&gt; 'PG.Q.Value' found in 6 out of 6 assay(s)</span></span>
<span><span class="co">#&gt; 'Q.Value' found in 6 out of 6 assay(s)</span></span>
<span><span class="co">#&gt; Warning: 'experiments' dropped; see 'drops()'</span></span>
<span><span class="co">#&gt; harmonizing input:</span></span>
<span><span class="co">#&gt;   removing 6 sampleMap rows not in names(experiments)</span></span></code></pre></div>
<p>We have 0 quantification values which are exactly zero here. Where
these do exist, they should be replaced with NA, since mass spectrometry
is not capable of asserting that the protein had zero abundance in the
sample.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'peptides_fdr_cntrl'</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">QFeatures</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-missing-data.html" class="external-link">zeroIsNA</a></span><span class="op">(</span><span class="va">dia_qf</span><span class="op">[[</span><span class="st">'peptides_fdr_cntrl'</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p>Our column names include redundant information (‘HYE_DIA’)</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">dia_qf</span><span class="op">)</span><span class="op">[[</span><span class="st">"peptides_fdr_cntrl"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">#&gt; [1] "HYE_DIA_A1" "HYE_DIA_A2" "HYE_DIA_A3" "HYE_DIA_B1" "HYE_DIA_B2"</span></span>
<span><span class="co">#&gt; [6] "HYE_DIA_B3"</span></span></code></pre></div>
<p>Below, we update the column names to remove the redundant information
and then update the sample names in the QFeatures</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">dia_qf</span><span class="op">)</span><span class="op">[[</span><span class="st">"peptides_fdr_cntrl"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">'HYE_DIA_'</span>, <span class="st">''</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">dia_qf</span><span class="op">)</span><span class="op">[[</span><span class="st">"peptides_fdr_cntrl"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">dia_qf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://github.com/waldronlab/MultiAssayExperiment/reference/MultiAssayExperiment-helpers.html" class="external-link">renamePrimary</a></span><span class="op">(</span><span class="va">dia_qf</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">dia_qf</span><span class="op">)</span><span class="op">[[</span><span class="st">"peptides_fdr_cntrl"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p>We need to populate the <code>colData</code>, which is currently
empty</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">dia_qf</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; data frame with 0 columns and 6 rows</span></span></code></pre></div>
<p>Here, we can create our <code>colData</code> from the sample names.
In other cases, you may need to read in a separate experimental design
file.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu">colData</span><span class="op">(</span><span class="va">dia_qf</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>quantCols<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">dia_qf</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/separate.html" class="external-link">separate</a></span><span class="op">(</span><span class="va">quantCols</span>, sep<span class="op">=</span><span class="st">''</span>, into<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="cn">NA</span>, <span class="st">'condition'</span>, <span class="st">'replicate'</span><span class="op">)</span>, remove <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">column_to_rownames</a></span><span class="op">(</span><span class="st">'quantCols'</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">dia_qf</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;    condition replicate</span></span>
<span><span class="co">#&gt; A1         A         1</span></span>
<span><span class="co">#&gt; A2         A         2</span></span>
<span><span class="co">#&gt; A3         A         3</span></span>
<span><span class="co">#&gt; B1         B         1</span></span>
<span><span class="co">#&gt; B2         B         2</span></span>
<span><span class="co">#&gt; B3         B         3</span></span></code></pre></div>
<p>Adding the <code>colData</code> to the peptide-level data too.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">colData</span><span class="op">(</span><span class="va">dia_qf</span><span class="op">[[</span><span class="st">'peptides_fdr_cntrl'</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">colData</span><span class="op">(</span><span class="va">dia_qf</span><span class="op">)</span></span></code></pre></div>
<p>We first perform routine filtering to remove PSMs that:</p>
<ul>
<li>Could originate from contaminants. See
<code><a href="../reference/filter_features_pd_dda.html">?filter_features_pd_dda</a></code> for further details, including the
removal of ‘associated cRAP’.</li>
<li>Don’t have a unique master protein.</li>
</ul>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dia_qf</span><span class="op">[[</span><span class="st">'peptides_filtered'</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">biomasslmb</span><span class="fu">::</span><span class="fu"><a href="../reference/filter_features_diann.html">filter_features_diann</a></span><span class="op">(</span><span class="va">dia_qf</span><span class="op">[[</span><span class="st">'peptides_fdr_cntrl'</span><span class="op">]</span><span class="op">]</span>, contaminant_proteins<span class="op">=</span><span class="va">crap_accessions</span><span class="op">)</span></span>
<span><span class="co">#&gt; Filtering data...</span></span>
<span><span class="co">#&gt; 5009 features found from 454 master proteins =&gt; Input</span></span>
<span><span class="co">#&gt; 242 contaminant proteins supplied</span></span>
<span><span class="co">#&gt; 11 proteins identified as 'contaminant associated'</span></span>
<span><span class="co">#&gt; 4998 features found from 453 master proteins =&gt; contaminant features removed</span></span>
<span><span class="co">#&gt; 4998 features found from 453 master proteins =&gt; associated contaminant features removed</span></span>
<span><span class="co">#&gt; 4998 features found from 453 master proteins =&gt; features without a master protein removed</span></span>
<span><span class="co">#&gt; 4955 features found from 448 master proteins =&gt; features with non-unique master proteins removed</span></span>
<span><span class="co">#&gt; 4955 features found from 448 master proteins =&gt; features without quantification removed</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="check-normalisation">Check normalisation<a class="anchor" aria-label="anchor" href="#check-normalisation"></a>
</h2>
<p>Next, we plot the peptide intensity distribution to check they are
approximately the same using <code><a href="../reference/plot_quant.html">biomasslmb::plot_quant</a></code>. By
default, DIA-NN performs normalisation. See the <a href="https://github.com/vdemichev/DiaNN?tab=readme-ov-file#readme" class="external-link">documentation</a>
for details of the underlying assumptions for the normalisation and use
cases where this normalisation approach may not be appropriate, e.g
interactome proteomics.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Plot the peptide-level quantification distributions per sample</span></span>
<span><span class="fu">biomasslmb</span><span class="fu">::</span><span class="fu"><a href="../reference/plot_quant.html">plot_quant</a></span><span class="op">(</span><span class="va">dia_qf</span><span class="op">[[</span><span class="st">'peptides_filtered'</span><span class="op">]</span><span class="op">]</span>,</span>
<span>                       log2transform<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>                       method<span class="op">=</span><span class="st">'density'</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="../reference/theme_biomasslmb.html">theme_biomasslmb</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">'PSM abundance (log2)'</span><span class="op">)</span></span></code></pre></div>
<p><img src="LFQ_DIA_Peptide_QC_files/figure-html/unnamed-chunk-11-1.png" width="75%"></p>
<div class="section level3">
<h3 id="missing-values">Missing values<a class="anchor" aria-label="anchor" href="#missing-values"></a>
</h3>
<p>Below, we inspect the number of missing values per sample</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-missing-data.html" class="external-link">nNA</a></span><span class="op">(</span><span class="va">dia_qf</span><span class="op">[[</span><span class="st">'peptides_filtered'</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">$</span><span class="va">nNAcols</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span>digits<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">name</th>
<th align="right">nNA</th>
<th align="right">pNA</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">A1</td>
<td align="right">641</td>
<td align="right">0.13</td>
</tr>
<tr class="even">
<td align="left">A2</td>
<td align="right">736</td>
<td align="right">0.15</td>
</tr>
<tr class="odd">
<td align="left">A3</td>
<td align="right">1029</td>
<td align="right">0.21</td>
</tr>
<tr class="even">
<td align="left">B1</td>
<td align="right">674</td>
<td align="right">0.14</td>
</tr>
<tr class="odd">
<td align="left">B2</td>
<td align="right">597</td>
<td align="right">0.12</td>
</tr>
<tr class="even">
<td align="left">B3</td>
<td align="right">613</td>
<td align="right">0.12</td>
</tr>
</tbody>
</table>
<p>Here, the samples have between 12 - 21 % missing values. This is
fairly typical of DIA data at the precursor level.</p>
<p>Next, we inspect the most common patterns for the missing values.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_missing_upset.html">plot_missing_upset</a></span><span class="op">(</span><span class="va">dia_qf</span>, i<span class="op">=</span><span class="st">'peptides_filtered'</span><span class="op">)</span></span></code></pre></div>
<p><img src="LFQ_DIA_Peptide_QC_files/figure-html/unnamed-chunk-13-1.png" width="768">
The missingness is frequently observed for all 3 replicates of a single
condition, but also commonly just for a single sample.</p>
</div>
<div class="section level3">
<h3 id="summarising-to-protein-level-abundances">Summarising to protein-level abundances<a class="anchor" aria-label="anchor" href="#summarising-to-protein-level-abundances"></a>
</h3>
<p>Now that we have inspected the peptide-level quantification and
filtered the peptides, we can summarise the peptide-level quantification
to protein-level abundances.</p>
<p>Since DIA typically has too many missing values to exclude peptides
with missing values, the <code><a href="https://rdrr.io/pkg/MsCoreUtils/man/robustSummary.html" class="external-link">MsCoreUtils::robustSummary</a></code> method
should preferred for summarisation to protein-level abundance. With
<code>robustSummary</code>, we do not need to remove all PSMs with
missing values since the summarisation algorithm deals with them
appropriately <span class="citation">(Sticker et al. 2020)</span>..
However, we still don’t want to retain PSMs with too many missing
values, since these will not be very informative in estimating the
protein-level quantification. Here, we will retain PSMs with at most 4/6
missing values</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dia_qf</span><span class="op">[[</span><span class="st">'peptides_filtered_missing'</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">QFeatures</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">filterNA</a></span><span class="op">(</span></span>
<span>  <span class="va">dia_qf</span><span class="op">[[</span><span class="st">'peptides_filtered'</span><span class="op">]</span><span class="op">]</span>, <span class="fl">4</span><span class="op">/</span><span class="fl">6</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">biomasslmb</span><span class="fu">:::</span><span class="fu">message_parse</span><span class="op">(</span><span class="fu">rowData</span><span class="op">(</span><span class="va">dia_qf</span><span class="op">[[</span><span class="st">'peptides_filtered_missing'</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>,</span>
<span>                           <span class="st">'Protein.Group'</span>,</span>
<span>                           <span class="st">"Removing peptides with &gt; 4/6 missing values"</span><span class="op">)</span></span>
<span><span class="co">#&gt; 4863 features found from 441 master proteins =&gt; Removing peptides with &gt; 4/6 missing values</span></span></code></pre></div>
<p>Next, we will remove PSMs for proteins with fewer than 2 PSMs. This
is a common filter in proteomics to ensure the protein-level
quantifications are derived from at least two independent observations.
In some cases, for example phosphoproteomics, or where this filter
appears to be too stringent, it may be appropriate to skip it.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">min_psms</span> <span class="op">&lt;-</span> <span class="fl">2</span></span>
<span> </span>
<span><span class="va">dia_qf</span><span class="op">[[</span><span class="st">'peptides_filtered_forRobust'</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">biomasslmb</span><span class="fu">::</span><span class="fu"><a href="../reference/filter_features_per_protein.html">filter_features_per_protein</a></span><span class="op">(</span></span>
<span>  <span class="va">dia_qf</span><span class="op">[[</span><span class="st">'peptides_filtered_missing'</span><span class="op">]</span><span class="op">]</span>, min_features <span class="op">=</span> <span class="va">min_psms</span>, master_protein_col<span class="op">=</span><span class="st">'Protein.Group'</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">biomasslmb</span><span class="fu">:::</span><span class="fu">message_parse</span><span class="op">(</span><span class="fu">rowData</span><span class="op">(</span><span class="va">dia_qf</span><span class="op">[[</span><span class="st">'peptides_filtered_forRobust'</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>,</span>
<span>                           <span class="st">'Protein.Group'</span>,</span>
<span>                           <span class="st">"Removing peptides for proteins with &lt; 2 peptides"</span><span class="op">)</span></span>
<span><span class="co">#&gt; 4829 features found from 407 master proteins =&gt; Removing peptides for proteins with &lt; 2 peptides</span></span></code></pre></div>
<p>Now we can summarise with <code>robustSummary</code></p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Aggregate to protein-level abundances (using QFeatures function)</span></span>
<span><span class="va">dia_qf</span> <span class="op">&lt;-</span> <span class="fu">QFeatures</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">aggregateFeatures</a></span><span class="op">(</span><span class="va">dia_qf</span>, </span>
<span>                                       i <span class="op">=</span> <span class="st">"peptides_filtered_forRobust"</span>, </span>
<span>                                       fcol <span class="op">=</span> <span class="st">"Protein.Group"</span>,</span>
<span>                                       name <span class="op">=</span> <span class="st">"protein"</span>,</span>
<span>                                       fun <span class="op">=</span> <span class="fu">MsCoreUtils</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/MsCoreUtils/man/robustSummary.html" class="external-link">robustSummary</a></span>,</span>
<span>                                       maxit<span class="op">=</span><span class="fl">10000</span><span class="op">)</span> <span class="co"># ensure sufficient iterations for convergance</span></span>
<span><span class="co">#&gt; Your quantitative data contain missing values. Please read the relevant</span></span>
<span><span class="co">#&gt; section(s) in the aggregateFeatures manual page regarding the effects</span></span>
<span><span class="co">#&gt; of missing values on data aggregation.</span></span>
<span></span>
<span><span class="fu">biomasslmb</span><span class="fu">:::</span><span class="fu">message_parse</span><span class="op">(</span><span class="fu">rowData</span><span class="op">(</span><span class="va">dia_qf</span><span class="op">[[</span><span class="st">'protein'</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>,</span>
<span>                           <span class="st">'Protein.Group'</span>,</span>
<span>                           <span class="st">"Summarised to proteins"</span><span class="op">)</span></span>
<span><span class="co">#&gt; 407 features found from 407 master proteins =&gt; Summarised to proteins</span></span></code></pre></div>
<p>Prior to summaristaion, we removed PSMs from proteins with fewer than
2 PSMs. However, since we left in PSMs with missing values, it’s
possible for some protein-level abundances to be derived from just a
single PSM. We can use the <code>get_protein_no_quant_mask</code> from
<code>biomasslmb</code> to identify where the protein abundances will be
derived from fewer than <code>n</code> features (PSMs). We can then give
this mask to <code>mask_protein_level_quant</code> to replace these
quantification values with NA.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># plot = TRUE means we will also get a plot of the number of proteins quantified in each sample</span></span>
<span><span class="va">protein_retain_mask</span> <span class="op">&lt;-</span> <span class="fu">biomasslmb</span><span class="fu">::</span><span class="fu"><a href="../reference/get_protein_no_quant_mask.html">get_protein_no_quant_mask</a></span><span class="op">(</span></span>
<span>  <span class="va">dia_qf</span><span class="op">[[</span><span class="st">'peptides_filtered_forRobust'</span><span class="op">]</span><span class="op">]</span>, min_features<span class="op">=</span><span class="fl">2</span>, plot<span class="op">=</span><span class="cn">TRUE</span>, master_protein_col<span class="op">=</span><span class="st">'Protein.Group'</span><span class="op">)</span> </span></code></pre></div>
<p><img src="LFQ_DIA_Peptide_QC_files/figure-html/unnamed-chunk-17-1.png" width="50%"></p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">dia_qf</span><span class="op">[[</span><span class="st">'protein'</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">biomasslmb</span><span class="fu">::</span><span class="fu"><a href="../reference/mask_protein_level_quant.html">mask_protein_level_quant</a></span><span class="op">(</span></span>
<span>  <span class="va">dia_qf</span><span class="op">[[</span><span class="st">'protein'</span><span class="op">]</span><span class="op">]</span>, <span class="va">protein_retain_mask</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="re-inspecting-missing-values-at-protein-level">Re-inspecting missing values at protein-level<a class="anchor" aria-label="anchor" href="#re-inspecting-missing-values-at-protein-level"></a>
</h3>
<p>Now that we have protein-level abundances, we would like to
re-inspect the missing values. Overall, we have 8.722 % missing values,
with at most 10.565 % missing values in any given sample. The most
common missingness pattern is for the protein to be missing from all 3
samples for a single condition.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-missing-data.html" class="external-link">nNA</a></span><span class="op">(</span><span class="va">dia_qf</span><span class="op">[[</span><span class="st">'protein'</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">$</span><span class="va">nNAcols</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span>digits<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">name</th>
<th align="right">nNA</th>
<th align="right">pNA</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">A1</td>
<td align="right">25</td>
<td align="right">0.06</td>
</tr>
<tr class="even">
<td align="left">A2</td>
<td align="right">31</td>
<td align="right">0.08</td>
</tr>
<tr class="odd">
<td align="left">A3</td>
<td align="right">43</td>
<td align="right">0.11</td>
</tr>
<tr class="even">
<td align="left">B1</td>
<td align="right">42</td>
<td align="right">0.10</td>
</tr>
<tr class="odd">
<td align="left">B2</td>
<td align="right">38</td>
<td align="right">0.09</td>
</tr>
<tr class="even">
<td align="left">B3</td>
<td align="right">34</td>
<td align="right">0.08</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu"><a href="../reference/plot_missing_upset.html">plot_missing_upset</a></span><span class="op">(</span><span class="va">dia_qf</span>, i<span class="op">=</span><span class="st">'protein'</span><span class="op">)</span></span></code></pre></div>
<p><img src="LFQ_DIA_Peptide_QC_files/figure-html/unnamed-chunk-18-1.png" width="768"></p>
</div>
<div class="section level3">
<h3 id="inspecting-the-number-of-peptides-and-proteins-through-the-processing-steps">Inspecting the number of peptides and proteins through the
processing steps<a class="anchor" aria-label="anchor" href="#inspecting-the-number-of-peptides-and-proteins-through-the-processing-steps"></a>
</h3>
<p>Now that we have performed all the filtering steps and summarisation
to protein-level abundances, it’s helpful to visualise how many
peptides/proteins were retained at each level of the processing. We can
use the <code><a href="../reference/get_samples_present.html">biomasslmb::get_samples_present</a></code> and
<code><a href="../reference/plot_samples_present.html">biomasslmb::plot_samples_present</a></code> functions for this. First
though, we need to decide which ‘experiments’ we want to plot and define
a named character vector since the QFeatures names are not sufficiently
clear by themselves</p>
<p>Below, we inspect the experiment names.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">dia_qf</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "peptides_fdr_cntrl"          "peptides_filtered"          </span></span>
<span><span class="co">#&gt; [3] "peptides_filtered_missing"   "peptides_filtered_forRobust"</span></span>
<span><span class="co">#&gt; [5] "protein"</span></span></code></pre></div>
<p>In this case, we want to plot all the experiments.</p>
<div class="section level4">
<h4 id="samples-per-psm">Samples per PSM<a class="anchor" aria-label="anchor" href="#samples-per-psm"></a>
</h4>
<p>We’ll start by inspecting the number of PSMs in each experiment. We
therefore define a named character vector all the PSM-level experiments,
exlcuding <code>psms_filtered_norm</code>. We set the row variables to
be <code>Precursor.Id</code> so that we count the number of unique
peptides (precursors).</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span></span>
<span></span>
<span><span class="va">rename_cols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'Peptides passing FDR thresholds'</span> <span class="op">=</span> <span class="st">'peptides_fdr_cntrl'</span> ,</span>
<span>                 <span class="st">'Quantified, contaminants removed'</span> <span class="op">=</span> <span class="st">'peptides_filtered'</span>,</span>
<span>                 <span class="st">'&lt;= 4/6 missing values'</span> <span class="op">=</span> <span class="st">'peptides_filtered_missing'</span>,</span>
<span>                 <span class="st">'&gt;1 peptides per protein'</span> <span class="op">=</span> <span class="st">'peptides_filtered_forRobust'</span><span class="op">)</span></span>
<span></span>
<span><span class="va">rowvars</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'Precursor.Id'</span><span class="op">)</span></span>
<span></span>
<span><span class="va">samples_present</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_samples_present.html">get_samples_present</a></span><span class="op">(</span><span class="va">dia_qf</span><span class="op">[</span>,,<span class="fu"><a href="https://rdrr.io/r/base/unname.html" class="external-link">unname</a></span><span class="op">(</span><span class="va">rename_cols</span><span class="op">)</span><span class="op">]</span>, <span class="va">rowvars</span>, <span class="va">rename_cols</span><span class="op">)</span></span>
<span><span class="co">#&gt; harmonizing input:</span></span>
<span><span class="co">#&gt;   removing 6 sampleMap rows not in names(experiments)</span></span>
<span><span class="fu"><a href="../reference/plot_samples_present.html">plot_samples_present</a></span><span class="op">(</span><span class="va">samples_present</span>, <span class="va">rowvars</span>, breaks<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">10</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">'PSM'</span><span class="op">)</span></span>
<span><span class="co">#&gt; Scale for <span style="color: #00BB00;">fill</span> is already present.</span></span>
<span><span class="co">#&gt; Adding another scale for <span style="color: #00BB00;">fill</span>, which will replace the existing scale.</span></span></code></pre></div>
<div class="figure">
<img src="LFQ_DIA_Peptide_QC_files/figure-html/unnamed-chunk-20-1.png" alt="Samples quantified for each PSM at each level of processing" width="75%"><p class="caption">
Samples quantified for each PSM at each level of processing
</p>
</div>
</div>
<div class="section level4">
<h4 id="samples-per-protein">Samples per Protein<a class="anchor" aria-label="anchor" href="#samples-per-protein"></a>
</h4>
<p>Next, we’ll use the same functions to inspect the number of proteins
at each level of processing. We need to supply an updated named
character vector to include the <code>protein</code> experiment and set
the row variables to be just the <code>Master.Protein.Accesions</code>
column.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">rename_cols_prot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">rename_cols</span>, <span class="st">'Protein'</span><span class="op">=</span><span class="st">'protein'</span><span class="op">)</span></span>
<span></span>
<span><span class="va">rowvars_prot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'Protein.Group'</span><span class="op">)</span></span>
<span></span>
<span><span class="va">samples_present</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_samples_present.html">get_samples_present</a></span><span class="op">(</span><span class="va">dia_qf</span>, <span class="va">rowvars_prot</span>, <span class="va">rename_cols_prot</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plot_samples_present.html">plot_samples_present</a></span><span class="op">(</span><span class="va">samples_present</span>, <span class="va">rowvars_prot</span>, breaks<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">10</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Scale for <span style="color: #00BB00;">fill</span> is already present.</span></span>
<span><span class="co">#&gt; Adding another scale for <span style="color: #00BB00;">fill</span>, which will replace the existing scale.</span></span></code></pre></div>
<div class="figure">
<img src="LFQ_DIA_Peptide_QC_files/figure-html/unnamed-chunk-21-1.png" alt="Samples quantified for each protein at each level of processing" width="75%"><p class="caption">
Samples quantified for each protein at each level of processing
</p>
</div>
<p>From these two plots, we can see that the filtering to ensure that
all proteins have &gt;1 peptides removed only a few peptides, but more
proteins. Whether this is appropriate will depend on your data in
hand.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.4.3 (2025-02-28)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 24.04.2 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] dplyr_1.1.4                 tidyr_1.3.1                </span></span>
<span><span class="co">#&gt;  [3] ggplot2_3.5.1               biomasslmb_0.0.2           </span></span>
<span><span class="co">#&gt;  [5] QFeatures_1.16.0            MultiAssayExperiment_1.32.0</span></span>
<span><span class="co">#&gt;  [7] SummarizedExperiment_1.36.0 Biobase_2.66.0             </span></span>
<span><span class="co">#&gt;  [9] GenomicRanges_1.58.0        GenomeInfoDb_1.42.3        </span></span>
<span><span class="co">#&gt; [11] IRanges_2.40.1              S4Vectors_0.44.0           </span></span>
<span><span class="co">#&gt; [13] BiocGenerics_0.52.0         MatrixGenerics_1.18.1      </span></span>
<span><span class="co">#&gt; [15] matrixStats_1.5.0          </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] DBI_1.2.3               gridExtra_2.3           rlang_1.1.5            </span></span>
<span><span class="co">#&gt;  [4] magrittr_2.0.3          clue_0.3-66             compiler_4.4.3         </span></span>
<span><span class="co">#&gt;  [7] RSQLite_2.3.9           png_0.1-8               systemfonts_1.2.1      </span></span>
<span><span class="co">#&gt; [10] vctrs_0.6.5             reshape2_1.4.4          stringr_1.5.1          </span></span>
<span><span class="co">#&gt; [13] ProtGenerics_1.38.0     pkgconfig_2.0.3         crayon_1.5.3           </span></span>
<span><span class="co">#&gt; [16] fastmap_1.2.0           XVector_0.46.0          labeling_0.4.3         </span></span>
<span><span class="co">#&gt; [19] rmarkdown_2.29          UCSC.utils_1.2.0        visdat_0.6.0           </span></span>
<span><span class="co">#&gt; [22] ragg_1.3.3              UpSetR_1.4.0            purrr_1.0.4            </span></span>
<span><span class="co">#&gt; [25] bit_4.6.0               xfun_0.51               zlibbioc_1.52.0        </span></span>
<span><span class="co">#&gt; [28] cachem_1.1.0            jsonlite_2.0.0          blob_1.2.4             </span></span>
<span><span class="co">#&gt; [31] DelayedArray_0.32.0     cluster_2.1.8           R6_2.6.1               </span></span>
<span><span class="co">#&gt; [34] bslib_0.9.0             stringi_1.8.7           genefilter_1.88.0      </span></span>
<span><span class="co">#&gt; [37] jquerylib_0.1.4         Rcpp_1.0.14             knitr_1.50             </span></span>
<span><span class="co">#&gt; [40] usethis_3.1.0           BiocBaseUtils_1.8.0     Matrix_1.7-2           </span></span>
<span><span class="co">#&gt; [43] splines_4.4.3           igraph_2.1.4            tidyselect_1.2.1       </span></span>
<span><span class="co">#&gt; [46] abind_1.4-8             yaml_2.3.10             lattice_0.22-6         </span></span>
<span><span class="co">#&gt; [49] tibble_3.2.1            plyr_1.8.9              withr_3.0.2            </span></span>
<span><span class="co">#&gt; [52] KEGGREST_1.46.0         evaluate_1.0.3          desc_1.4.3             </span></span>
<span><span class="co">#&gt; [55] survival_3.8-3          Biostrings_2.74.1       pillar_1.10.1          </span></span>
<span><span class="co">#&gt; [58] corrplot_0.95           generics_0.1.3          rprojroot_2.0.4        </span></span>
<span><span class="co">#&gt; [61] munsell_0.5.1           scales_1.3.0            xtable_1.8-4           </span></span>
<span><span class="co">#&gt; [64] glue_1.8.0              lazyeval_0.2.2          tools_4.4.3            </span></span>
<span><span class="co">#&gt; [67] robustbase_0.99-4-1     annotate_1.84.0         fs_1.6.5               </span></span>
<span><span class="co">#&gt; [70] XML_3.99-0.18           grid_4.4.3              MsCoreUtils_1.18.0     </span></span>
<span><span class="co">#&gt; [73] AnnotationDbi_1.68.0    colorspace_2.1-1        GenomeInfoDbData_1.2.13</span></span>
<span><span class="co">#&gt; [76] naniar_1.1.0            cli_3.6.4               textshaping_1.0.0      </span></span>
<span><span class="co">#&gt; [79] S4Arrays_1.6.0          AnnotationFilter_1.30.0 gtable_0.3.6           </span></span>
<span><span class="co">#&gt; [82] DEoptimR_1.1-3-1        sass_0.4.9              digest_0.6.37          </span></span>
<span><span class="co">#&gt; [85] SparseArray_1.6.2       htmlwidgets_1.6.4       farver_2.1.2           </span></span>
<span><span class="co">#&gt; [88] memoise_2.0.1           htmltools_0.5.8.1       pkgdown_2.1.1          </span></span>
<span><span class="co">#&gt; [91] lifecycle_1.0.4         httr_1.4.7              bit64_4.6.0-1          </span></span>
<span><span class="co">#&gt; [94] MASS_7.3-64</span></span></code></pre></div>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-http://zotero.org/users/5634351/items/FZN3QTTZ" class="csl-entry">
Sticker, Adriaan, Ludger Goeminne, Lennart Martens, and Lieven Clement.
2020. <span>“Robust Summarization and Inference in Proteome-wide
Label-free Quantification.”</span> <em>Molecular &amp; cellular
proteomics: MCP</em> 19 (7): 1209–19. <a href="https://doi.org/10.1074/mcp.RA119.001624" class="external-link">https://doi.org/10.1074/mcp.RA119.001624</a>.
</div>
</div>
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Tom Smith.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
