<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>LFQ-DDA QC Peptide-level quantification and summarisation to protein-level abundance • biomasslmb</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="LFQ-DDA QC Peptide-level quantification and summarisation to protein-level abundance">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">biomasslmb</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.4</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/LFQ_DDA_Peptide_QC_Summarisation.html">LFQ-DDA QC Peptide-level quantification and summarisation to protein-level abundance</a></li>
    <li><a class="dropdown-item" href="../articles/LFQ_DIA_Peptide_QC.html">LFQ-DIA QC Peptide-level quantification and summarisation to protein-level abundance</a></li>
    <li><a class="dropdown-item" href="../articles/TMT_PSM_QC_Summarisation.html">TMT QC PSM-level quantification and summarisation to protein-level abundance</a></li>
    <li><a class="dropdown-item" href="../articles/TMT_summarisation_methods.html">TMT - comparing protein summarisation approaches</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>LFQ-DDA QC Peptide-level quantification and summarisation to protein-level abundance</h1>
                        <h4 data-toc-skip class="author">Tom Smith</h4>
            
            <h4 data-toc-skip class="date">2025-07-10</h4>
      

      <div class="d-none name"><code>LFQ_DDA_Peptide_QC_Summarisation.Rmd</code></div>
    </div>

    
    
<p>Label-Free Quantification (LFQ) is the simplest form of quantitative
proteomics, in which different samples are quantified in separate MS
runs. Quantification is either performed by Data-Dependent Aquisition
(DDA), where the Mass Spectrometer triggers fragmentation of ions within
a given m/z range with the aim being to focus attention of individual
peptides separately, or Data-Independent Aquisition (DIA), where a much
wider m/z range is used and a mix of peptides are co-fragmented and
quantified simultaneously by deconvoluting the resultant complex
spectra. Here, we will focus on just LFQ-DDA.</p>
<p>Since each sample is run separately, different peptides will be
quantified in each sample and the peptide intensities may not be
directly comparable between samples. The common solution to the higher
burden of missing values is to use the ‘match-between-runs’ (Cox et
al. 2014[UPDATE REF!!!]), or the functionally equivalent ‘Minora’
algorithm employed by Proteome Discoverer (PD). These algorithms use the
observed retention times of MS1 ions which were successfully spectrum
matched in one sample to identify the likely peptide sequence of MS1
ions that could not be spectrum matched in another sample. However, even
with these algorithms, DDA LFQ will still typically have many more
missing values than labelled LFQ, e.g TMT.</p>
<p>Despite the pitfalls of LFQ, the data analysis is still relatively
straightforward, though there are steps that need some careful
consideration and quality control assessment.</p>
<div class="section level2">
<h2 id="load-required-packages">Load required packages<a class="anchor" aria-label="anchor" href="#load-required-packages"></a>
</h2>
<p>To clarify which functionality is provided by which package, we will
use <code>package::function</code>. For your own code, there is no need
to specify the package unless you want to maintain this clarity.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/RforMassSpectrometry/QFeatures" class="external-link">QFeatures</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://lmb-mass-spec-compbio.github.io/biomasslmb/">biomasslmb</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org" class="external-link">tidyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="defining-the-contaminant-proteins">Defining the contaminant proteins<a class="anchor" aria-label="anchor" href="#defining-the-contaminant-proteins"></a>
</h2>
<p>We need to remove contaminant proteins. These were defined here using
the cRAP database. Below, we parse the contaminants fasta to extract the
IDs for the proteins in both ‘cRAP’ format and Uniprot IDs.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">crap_fasta_inf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span></span>
<span>  <span class="st">"extdata"</span>, <span class="st">"cRAP_20190401.fasta.gz"</span>, </span>
<span>  package <span class="op">=</span> <span class="st">"biomasslmb"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Extract the protein IDs associated with each cRAP protein</span></span>
<span><span class="va">crap_accessions</span> <span class="op">&lt;-</span> <span class="fu">biomasslmb</span><span class="fu">::</span><span class="fu"><a href="../reference/get_crap_fasta_accessions.html">get_crap_fasta_accessions</a></span><span class="op">(</span><span class="va">crap_fasta_inf</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">crap_accessions</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "cRAP001" "P00330"  "cRAP002" "P02768"  "cRAP003" "P00883"</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="read-in-input-data">Read in input data<a class="anchor" aria-label="anchor" href="#read-in-input-data"></a>
</h2>
<p>We start by reading in quantitative proteomics data into a
<code>QFeatures</code> object, which is the standard Bioconductor object
for holding quantitative proteomics data. See <a href="https://www.bioconductor.org/packages/release/bioc/html/QFeatures.html" class="external-link">here</a>
for documentation about the <code>QFeatures</code> object.In this case,
we are not adding any experimental details to the <code>QFeatures</code>
object, so the <code>ColData</code> will be empty.</p>
<p><code>??</code> is a data set available from the
<code>biomasslmb</code> package</p>
<p><code>lfq_dda_pd_PeptideGroups.txt</code> is a file available from
the <code>biomasslmb</code> package containing the Peptide-level output
from Proteome Discoverer (PD) for an experiment with 6 samples. It is a
truncated file containing the precursors peptides quantification for
just 500 proteins.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pep_inf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span></span>
<span>  <span class="st">"extdata"</span>, <span class="st">"lfq_dda_pd_PeptideGroups.txt"</span>, </span>
<span>  package <span class="op">=</span> <span class="st">"biomasslmb"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Inspecting the column names for the abundances, we see the names are
not particularly useful. The <code>Abundance.FXX.na.Sample.</code> part
is irrelevant and the replicate number is not included</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">infdf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.delim</a></span><span class="op">(</span><span class="va">pep_inf</span><span class="op">)</span></span>
<span></span>
<span><span class="va">abundance_cols_ix</span> <span class="op">&lt;-</span><span class="fl">18</span><span class="op">:</span><span class="fl">23</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">infdf</span><span class="op">)</span><span class="op">[</span><span class="va">abundance_cols_ix</span><span class="op">]</span></span>
<span><span class="co">#&gt; [1] "Abundance.F12.na.Sample.A" "Abundance.F13.na.Sample.A"</span></span>
<span><span class="co">#&gt; [3] "Abundance.F14.na.Sample.A" "Abundance.F15.na.Sample.B"</span></span>
<span><span class="co">#&gt; [5] "Abundance.F16.na.Sample.B" "Abundance.F17.na.Sample.B"</span></span></code></pre></div>
<p>Below, we update the column names for the abundance columns</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Clean up sample names</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">infdf</span><span class="op">)</span><span class="op">[</span><span class="va">abundance_cols_ix</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">'Abundance.F\\d+.na.Sample.'</span>, <span class="st">''</span>,</span>
<span>                                        <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">infdf</span><span class="op">)</span><span class="op">[</span><span class="va">abundance_cols_ix</span><span class="op">]</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, times<span class="op">=</span><span class="fl">2</span><span class="op">)</span>, sep<span class="op">=</span><span class="st">'_'</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">infdf</span><span class="op">)</span><span class="op">[</span><span class="va">abundance_cols_ix</span><span class="op">]</span></span>
<span><span class="co">#&gt; [1] "A_1" "A_2" "A_3" "B_1" "B_2" "B_3"</span></span></code></pre></div>
<p>Before we read in the data, we also want to define the experimental
conditions for each sample. We can then provide the</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">exp_design</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>quantCols<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">infdf</span><span class="op">)</span><span class="op">[</span><span class="va">abundance_cols_ix</span><span class="op">]</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/separate.html" class="external-link">separate</a></span><span class="op">(</span><span class="va">quantCols</span>, sep<span class="op">=</span><span class="st">'_'</span>, into<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'condition'</span>, <span class="st">'replicate'</span><span class="op">)</span>, remove <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> </span>
<span></span>
<span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="va">exp_design</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">quantCols</th>
<th align="left">condition</th>
<th align="left">replicate</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">A_1</td>
<td align="left">A</td>
<td align="left">1</td>
</tr>
<tr class="even">
<td align="left">A_2</td>
<td align="left">A</td>
<td align="left">2</td>
</tr>
<tr class="odd">
<td align="left">A_3</td>
<td align="left">A</td>
<td align="left">3</td>
</tr>
<tr class="even">
<td align="left">B_1</td>
<td align="left">B</td>
<td align="left">1</td>
</tr>
<tr class="odd">
<td align="left">B_2</td>
<td align="left">B</td>
<td align="left">2</td>
</tr>
<tr class="even">
<td align="left">B_3</td>
<td align="left">B</td>
<td align="left">3</td>
</tr>
</tbody>
</table>
<p>Now we can read the data into a <code>QFeatures</code> object</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Read in PSM-level quantification from TMT experiment (using QFeatures function)</span></span>
<span><span class="va">lfq_qf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/readQFeatures.html" class="external-link">readQFeatures</a></span><span class="op">(</span>assayData <span class="op">=</span> <span class="va">infdf</span>,</span>
<span>                        quantCols <span class="op">=</span> <span class="va">abundance_cols_ix</span>, </span>
<span>                        colData <span class="op">=</span> <span class="va">exp_design</span>,</span>
<span>                        name <span class="op">=</span> <span class="st">"peptides_raw"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Checking arguments.</span></span>
<span><span class="co">#&gt; Loading data as a 'SummarizedExperiment' object.</span></span>
<span><span class="co">#&gt; Formatting sample annotations (colData).</span></span>
<span><span class="co">#&gt; Formatting data as a 'QFeatures' object.</span></span></code></pre></div>
<p>Adding the <code>colData</code> to the peptide-level data too.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">colData</span><span class="op">(</span><span class="va">lfq_qf</span><span class="op">[[</span><span class="st">'peptides_raw'</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">colData</span><span class="op">(</span><span class="va">lfq_qf</span><span class="op">)</span></span></code></pre></div>
<p>Extract the gene name and long-form protein name from the
descriptions.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'peptides_raw'</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/add_gene_long_protein_name_pd.html">add_gene_long_protein_name_pd</a></span><span class="op">(</span><span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'peptides_raw'</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p>It’s helpful to assess the retention time (RT) for the peptides and
the relationship with the delta mass for the precursor ion. Everything
looks OK here.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_rt_dist.html">plot_rt_dist</a></span><span class="op">(</span><span class="va">lfq_qf</span><span class="op">[[</span><span class="st">'peptides_raw'</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plot_rt_vs_delta.html">plot_rt_vs_delta</a></span><span class="op">(</span><span class="va">lfq_qf</span><span class="op">[[</span><span class="st">'peptides_raw'</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="LFQ_DDA_Peptide_QC_Summarisation_files/figure-html/unnamed-chunk-10-1.png" width="50%"><img src="LFQ_DDA_Peptide_QC_Summarisation_files/figure-html/unnamed-chunk-10-2.png" width="50%"></p>
</div>
<div class="section level2">
<h2 id="filter-peptides">Filter peptides<a class="anchor" aria-label="anchor" href="#filter-peptides"></a>
</h2>
<p>We first perform routine filtering to remove PSMs that:</p>
<ul>
<li>Could originate from contaminants. See <code>?filter_features</code>
for further details, including the removal of ‘associated cRAP’.</li>
<li>Don’t have a unique master protein</li>
<li>Don’t have any quantification values</li>
</ul>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Perform routine raw data filtering.</span></span>
<span><span class="co"># - Remove PSMs from contaminant proteins</span></span>
<span><span class="co"># - Remove PSMs where protein ID is empty or not unique</span></span>
<span></span>
<span><span class="va">lfq_qf</span><span class="op">[[</span><span class="st">'peptides_filtered'</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filter_features_pd_dda.html">filter_features_pd_dda</a></span><span class="op">(</span><span class="va">lfq_qf</span><span class="op">[[</span><span class="st">'peptides_raw'</span><span class="op">]</span><span class="op">]</span>,</span>
<span>                                                        protein_col<span class="op">=</span><span class="st">'Master.Protein.Accessions'</span>, <span class="co"># UPDATE WHEN RE-EXPORTED</span></span>
<span>                                                 contaminant_proteins<span class="op">=</span><span class="va">crap_accessions</span>,</span>
<span>                                                 filter_contaminant <span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>                                                 filter_associated_contaminant<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>                                                 unique_master<span class="op">=</span><span class="cn">FALSE</span>,</span>
<span>                                                 remove_no_quant <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Filtering data...</span></span>
<span><span class="co">#&gt; 3106 features found from 500 master proteins =&gt; Input</span></span>
<span><span class="co">#&gt; 242 contaminant proteins supplied</span></span>
<span><span class="co">#&gt; 15 proteins identified as 'contaminant associated'</span></span>
<span><span class="co">#&gt; 3091 features found from 498 master proteins =&gt; contaminant features removed</span></span>
<span><span class="co">#&gt; 3091 features found from 498 master proteins =&gt; associated contaminant features removed</span></span>
<span><span class="co">#&gt; 3091 features found from 498 master proteins =&gt; features without a master protein removed</span></span>
<span><span class="co">#&gt; 2746 features found from 461 master proteins =&gt; features without quantification removed</span></span></code></pre></div>
<p>We will also remove peptides that are not rank 1 according to the
search engine</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">lfq_qf</span> <span class="op">&lt;-</span> <span class="va">lfq_qf</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/filterFeatures.html" class="external-link">filterFeatures</a></span><span class="op">(</span><span class="op">~</span> <span class="va">Search.Engine.Rank.by.Search.Engine.Sequest.HT</span> <span class="op">==</span> <span class="fl">1</span>,</span>
<span>               i <span class="op">=</span> <span class="st">"peptides_filtered"</span><span class="op">)</span> </span>
<span><span class="co">#&gt; 'Search.Engine.Rank.by.Search.Engine.Sequest.HT' found in 2 out of 2 assay(s).</span></span>
<span></span>
<span><span class="fu">biomasslmb</span><span class="fu">:::</span><span class="fu">message_parse</span><span class="op">(</span><span class="fu">rowData</span><span class="op">(</span><span class="va">lfq_qf</span><span class="op">[[</span><span class="st">'peptides_filtered'</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>,</span>
<span>                           <span class="st">'Master.Protein.Accessions'</span>,</span>
<span>                           <span class="st">"Removing peptides that are not rank 1"</span><span class="op">)</span></span>
<span><span class="co">#&gt; 2741 features found from 461 master proteins =&gt; Removing peptides that are not rank 1</span></span></code></pre></div>
<p>For the normalisation and summarisation, we need our quantification
values to be on a log-scale, so we log-transform below</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lfq_qf</span><span class="op">[[</span><span class="st">'peptides_filtered'</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">QFeatures</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-processing.html" class="external-link">logTransform</a></span><span class="op">(</span></span>
<span>  <span class="va">lfq_qf</span><span class="op">[[</span><span class="st">'peptides_filtered'</span><span class="op">]</span><span class="op">]</span>, base<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p>Next, we plot the peptide intensities.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Plot the peptide-level quantification distributions per sample</span></span>
<span><span class="fu"><a href="../reference/plot_quant.html">plot_quant</a></span><span class="op">(</span><span class="va">lfq_qf</span><span class="op">[[</span><span class="st">'peptides_filtered'</span><span class="op">]</span><span class="op">]</span>, log2transform<span class="op">=</span><span class="cn">FALSE</span>, method<span class="op">=</span><span class="st">'density'</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_colour_manual</a></span><span class="op">(</span>values<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="../reference/get_cat_palette.html">get_cat_palette</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span>, each<span class="op">=</span><span class="fl">3</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>linetype<span class="op">=</span><span class="va">sample</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_linetype_manual</a></span><span class="op">(</span>values<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, times<span class="op">=</span><span class="fl">3</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">'Peptide abundance (log2)'</span><span class="op">)</span></span></code></pre></div>
<p><img src="LFQ_DDA_Peptide_QC_Summarisation_files/figure-html/unnamed-chunk-14-1.png" width="50%"></p>
<div class="section level3">
<h3 id="normalise">Normalise<a class="anchor" aria-label="anchor" href="#normalise"></a>
</h3>
<p>Since the same amount of sample was injected for each sample, we want
to normalise to remove differences in the overall disribution of
abundances. We will thus perform ‘diff.median’ normalisation with
<code><a href="https://rdrr.io/pkg/BiocGenerics/man/normalize.html" class="external-link">QFeatures::normalize</a></code>.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lfq_qf</span><span class="op">[[</span><span class="st">'peptides_norm'</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">QFeatures</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/normalize.html" class="external-link">normalize</a></span><span class="op">(</span><span class="va">lfq_qf</span><span class="op">[[</span><span class="st">'peptides_filtered'</span><span class="op">]</span><span class="op">]</span>,</span>
<span>                                                  method<span class="op">=</span><span class="st">'diff.median'</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Plot the peptide-level quantification distributions per sample</span></span>
<span><span class="fu"><a href="../reference/plot_quant.html">plot_quant</a></span><span class="op">(</span><span class="va">lfq_qf</span><span class="op">[[</span><span class="st">'peptides_norm'</span><span class="op">]</span><span class="op">]</span>, log2transform<span class="op">=</span><span class="cn">FALSE</span>, method<span class="op">=</span><span class="st">'density'</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_colour_manual</a></span><span class="op">(</span>values<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="../reference/get_cat_palette.html">get_cat_palette</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span>, each<span class="op">=</span><span class="fl">3</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>linetype<span class="op">=</span><span class="va">sample</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_linetype_manual</a></span><span class="op">(</span>values<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, times<span class="op">=</span><span class="fl">3</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">'Peptide abundance (log2)'</span><span class="op">)</span></span></code></pre></div>
<p><img src="LFQ_DDA_Peptide_QC_Summarisation_files/figure-html/unnamed-chunk-16-1.png" width="50%"></p>
</div>
<div class="section level3">
<h3 id="summarising-to-protein-level-abundances">Summarising to protein-level abundances<a class="anchor" aria-label="anchor" href="#summarising-to-protein-level-abundances"></a>
</h3>
<p>For the summarisation to protein, we will use
<code>robustSummary</code>. We do not need to remove peptides with any
missing values. However, we still don’t want to retain peptides with too
many missing values, since these will not be very informative in
estimating the protein-level quantification.</p>
<p>Below, we inspect the most common patterns for the missing
values.</p>
<p>The missingness is most frequently consigned to all samples from one
condition and also commonly 2/3 replicates from one condition. It looks
like the missingness is far from missing at random (MAR) and is likely
missing due to falling below detection limits.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_missing_upset.html">plot_missing_upset</a></span><span class="op">(</span><span class="va">lfq_qf</span>, i<span class="op">=</span><span class="st">'peptides_norm'</span><span class="op">)</span></span></code></pre></div>
<p><img src="LFQ_DDA_Peptide_QC_Summarisation_files/figure-html/unnamed-chunk-17-1.png" width="768"></p>
<p>Since DDA LFQ typically has too many missing values to exclude
peptides with missing values, the
<code><a href="https://rdrr.io/pkg/MsCoreUtils/man/robustSummary.html" class="external-link">MsCoreUtils::robustSummary</a></code> method should preferred for
summarisation to protein-level abundance. With
<code>robustSummary</code>, we do not need to remove all PSMs with
missing values since the summarisation algorithm deals with them
appropriately <span class="citation">(Sticker et al. 2020)</span>..
However, we still don’t want to retain PSMs with too many missing
values, since these will not be very informative in estimating the
protein-level quantification. Here, we will retain PSMs with at most 4/6
missing values</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">lfq_qf</span><span class="op">[[</span><span class="st">'peptides_filtered_missing'</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">QFeatures</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">filterNA</a></span><span class="op">(</span></span>
<span>  <span class="va">lfq_qf</span><span class="op">[[</span><span class="st">'peptides_norm'</span><span class="op">]</span><span class="op">]</span>, <span class="fl">4</span><span class="op">/</span><span class="fl">6</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">biomasslmb</span><span class="fu">:::</span><span class="fu">message_parse</span><span class="op">(</span><span class="fu">rowData</span><span class="op">(</span><span class="va">lfq_qf</span><span class="op">[[</span><span class="st">'peptides_filtered_missing'</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>,</span>
<span>                           <span class="st">'Master.Protein.Accessions'</span>,</span>
<span>                           <span class="st">"Removing peptides with more than 4/6 missing values"</span><span class="op">)</span></span>
<span><span class="co">#&gt; 2704 features found from 457 master proteins =&gt; Removing peptides with more than 4/6 missing values</span></span></code></pre></div>
<p>Next, we remove peptides for proteins with fewer than 2 peptides.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">min_peps</span> <span class="op">&lt;-</span> <span class="fl">2</span></span>
<span><span class="va">lfq_qf</span><span class="op">[[</span><span class="st">'peptides_for_summarisation'</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filter_features_per_protein.html">filter_features_per_protein</a></span><span class="op">(</span></span>
<span>  <span class="va">lfq_qf</span><span class="op">[[</span><span class="st">'peptides_filtered_missing'</span><span class="op">]</span><span class="op">]</span>, min_features <span class="op">=</span> <span class="va">min_peps</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">biomasslmb</span><span class="fu">:::</span><span class="fu">message_parse</span><span class="op">(</span><span class="fu">rowData</span><span class="op">(</span><span class="va">lfq_qf</span><span class="op">[[</span><span class="st">'peptides_for_summarisation'</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>,</span>
<span>                           <span class="st">'Master.Protein.Accessions'</span>,</span>
<span>                           <span class="st">"Removing 'one-hit' wonders"</span><span class="op">)</span></span>
<span><span class="co">#&gt; 2586 features found from 339 master proteins =&gt; Removing 'one-hit' wonders</span></span></code></pre></div>
<p>Now we can summarise with <code>robustSummary</code></p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">42</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Aggregate to protein-level abundances (using QFeatures function)</span></span>
<span><span class="va">lfq_qf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">aggregateFeatures</a></span><span class="op">(</span><span class="va">lfq_qf</span>, </span>
<span>                            i <span class="op">=</span> <span class="st">"peptides_for_summarisation"</span>,</span>
<span>                            <span class="co">#i = "peptides_filtered_missing", </span></span>
<span>                            fcol <span class="op">=</span> <span class="st">"Master.Protein.Accessions"</span>,</span>
<span>                            name <span class="op">=</span> <span class="st">"protein"</span>,</span>
<span>                            fun <span class="op">=</span> <span class="fu">MsCoreUtils</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/MsCoreUtils/man/robustSummary.html" class="external-link">robustSummary</a></span>,</span>
<span>                            maxit<span class="op">=</span><span class="fl">10000</span><span class="op">)</span></span>
<span><span class="co">#&gt; Your quantitative and row data contain missing values. Please read the</span></span>
<span><span class="co">#&gt; relevant section(s) in the aggregateFeatures manual page regarding the</span></span>
<span><span class="co">#&gt; effects of missing values on data aggregation.</span></span></code></pre></div>
<p>Prior to summaristaion, we removed PSMs from proteins with fewer than
2 peptides. However, since we left in peptides with missing values, it’s
possible for some protein-level abundances to be derived from just a
single peptide. We can use the <code>get_protein_no_quant_mask</code>
from <code>biomasslmb</code> to identify where the protein abundances
will be derived from fewer than <code>n</code> features (peptides). We
can then give this mask to <code>mask_protein_level_quant</code> to
replace these quantification values with NA.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># plot = TRUE means we will also get a plot of the number of proteins quantified in each sample</span></span>
<span><span class="va">protein_retain_mask</span> <span class="op">&lt;-</span> <span class="fu">biomasslmb</span><span class="fu">::</span><span class="fu"><a href="../reference/get_protein_no_quant_mask.html">get_protein_no_quant_mask</a></span><span class="op">(</span></span>
<span>  <span class="va">lfq_qf</span><span class="op">[[</span><span class="st">'peptides_for_summarisation'</span><span class="op">]</span><span class="op">]</span>, min_features<span class="op">=</span><span class="va">min_peps</span>, plot<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">lfq_qf</span><span class="op">[[</span><span class="st">'protein'</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">biomasslmb</span><span class="fu">::</span><span class="fu"><a href="../reference/mask_protein_level_quant.html">mask_protein_level_quant</a></span><span class="op">(</span></span>
<span>  <span class="va">lfq_qf</span><span class="op">[[</span><span class="st">'protein'</span><span class="op">]</span><span class="op">]</span>, <span class="va">protein_retain_mask</span><span class="op">)</span></span></code></pre></div>
<p><img src="LFQ_DDA_Peptide_QC_Summarisation_files/figure-html/unnamed-chunk-21-1.png" width="50%"></p>
</div>
<div class="section level3">
<h3 id="re-inspecting-missing-values-at-protein-level">Re-inspecting missing values at protein-level<a class="anchor" aria-label="anchor" href="#re-inspecting-missing-values-at-protein-level"></a>
</h3>
<p>Now that we have protein-level abundances, we would like to
re-inspect the missing values. Overall, we have 5.064 % missing values,
with at most 10.565 % missing values in any given sample. The most
common missingness patterns are for the protein to be missing from 1,2
or 3 replicates from a single condition, suggesting the quantification
is missing due to detection limits.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu"><a href="../reference/plot_missing_upset.html">plot_missing_upset</a></span><span class="op">(</span><span class="va">lfq_qf</span>, i<span class="op">=</span><span class="st">'protein'</span> <span class="op">)</span></span></code></pre></div>
<p><img src="LFQ_DDA_Peptide_QC_Summarisation_files/figure-html/unnamed-chunk-22-1.png" width="768"></p>
</div>
<div class="section level3">
<h3 id="inspecting-the-number-of-peptides-and-proteins-through-the-processing-steps">Inspecting the number of peptides and proteins through the
processing steps<a class="anchor" aria-label="anchor" href="#inspecting-the-number-of-peptides-and-proteins-through-the-processing-steps"></a>
</h3>
<p>Now that we have performed all the filtering steps and summarisation
to protein-level abundances, it’s helpful to visualise how many
peptides/proteins were retained at each level of the processing. We can
use the <code><a href="../reference/get_samples_present.html">biomasslmb::get_samples_present</a></code> and
<code><a href="../reference/plot_samples_present.html">biomasslmb::plot_samples_present</a></code> functions for this. First
though, we need to decide which ‘experiments’ we want to plot and define
a named character vector since the QFeatures names are not sufficiently
clear by themselves</p>
<p>Below, we inspect the experiment names.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">lfq_qf</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "peptides_raw"               "peptides_filtered"         </span></span>
<span><span class="co">#&gt; [3] "peptides_norm"              "peptides_filtered_missing" </span></span>
<span><span class="co">#&gt; [5] "peptides_for_summarisation" "protein"</span></span></code></pre></div>
<p>In this case, we don’t want to plot the <code>peptides_norm</code>,
since that’s just the normalised quantification present in
<code>peptides_filtered</code> and no peptides were removed in that
step.</p>
<div class="section level4">
<h4 id="samples-per-peptide">Samples per peptide<a class="anchor" aria-label="anchor" href="#samples-per-peptide"></a>
</h4>
<p>We’ll start by inspecting the number of peptides in each experiment.
We therefore define a named character vector all the peptides-level
experiments, excluding <code>psms_filtered_norm</code>. We set the row
variables to be <code>Annotated.Sequence</code> and
<code>Modifications</code> so that we count the number of unique
peptides.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span></span>
<span><span class="va">rename_cols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'All peptides'</span> <span class="op">=</span> <span class="st">'peptides_raw'</span> ,</span>
<span>                 <span class="st">'Quantified, contaminants removed'</span> <span class="op">=</span> <span class="st">'peptides_filtered'</span>,</span>
<span>                 <span class="st">'No missing values'</span> <span class="op">=</span> <span class="st">'peptides_filtered_missing'</span>,</span>
<span>                 <span class="st">'&gt;1 PSM per protein'</span> <span class="op">=</span> <span class="st">'peptides_for_summarisation'</span><span class="op">)</span></span>
<span></span>
<span><span class="va">rowvars</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'Annotated.Sequence'</span>, <span class="st">'Modifications'</span><span class="op">)</span></span>
<span></span>
<span><span class="va">samples_present</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_samples_present.html">get_samples_present</a></span><span class="op">(</span><span class="va">lfq_qf</span><span class="op">[</span>,,<span class="fu"><a href="https://rdrr.io/r/base/unname.html" class="external-link">unname</a></span><span class="op">(</span><span class="va">rename_cols</span><span class="op">)</span><span class="op">]</span>, <span class="va">rowvars</span>, <span class="va">rename_cols</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: 'experiments' dropped; see 'drops()'</span></span>
<span><span class="co">#&gt; harmonizing input:</span></span>
<span><span class="co">#&gt;   removing 12 sampleMap rows not in names(experiments)</span></span>
<span><span class="co">#&gt; Warning in lifeCycle("longForm", package = "MultiAssayExperiment", title = "longFormat"): 'longFormat' is deprecated.</span></span>
<span><span class="co">#&gt;   Use 'longForm' instead.</span></span>
<span><span class="co">#&gt;   See help('longFormat-deprecated').</span></span>
<span><span class="fu"><a href="../reference/plot_samples_present.html">plot_samples_present</a></span><span class="op">(</span><span class="va">samples_present</span>, <span class="va">rowvars</span>, breaks<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">10</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">'PSM'</span><span class="op">)</span></span>
<span><span class="co">#&gt; Scale for <span style="color: #00BB00;">fill</span> is already present.</span></span>
<span><span class="co">#&gt; Adding another scale for <span style="color: #00BB00;">fill</span>, which will replace the existing scale.</span></span></code></pre></div>
<div class="figure">
<img src="LFQ_DDA_Peptide_QC_Summarisation_files/figure-html/unnamed-chunk-24-1.png" alt="Samples quantified for each PSM at each level of processing" width="75%"><p class="caption">
Samples quantified for each PSM at each level of processing
</p>
</div>
</div>
<div class="section level4">
<h4 id="samples-per-protein">Samples per Protein<a class="anchor" aria-label="anchor" href="#samples-per-protein"></a>
</h4>
<p>Next, we’ll use the same functions to inspect the number of proteins
at each level of processing. We need to supply an updated named
character vector to include the <code>protein</code> experiment and set
the row variables to be just the <code>Master.Protein.Accesions</code>
column.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">rename_cols_prot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">rename_cols</span>, <span class="st">'Protein'</span><span class="op">=</span><span class="st">'protein'</span><span class="op">)</span></span>
<span></span>
<span><span class="va">rowvars_prot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'Master.Protein.Accessions'</span><span class="op">)</span></span>
<span></span>
<span><span class="va">samples_present</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_samples_present.html">get_samples_present</a></span><span class="op">(</span><span class="va">lfq_qf</span>, <span class="va">rowvars_prot</span>, <span class="va">rename_cols_prot</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in lifeCycle("longForm", package = "MultiAssayExperiment", title = "longFormat"): 'longFormat' is deprecated.</span></span>
<span><span class="co">#&gt;   Use 'longForm' instead.</span></span>
<span><span class="co">#&gt;   See help('longFormat-deprecated').</span></span>
<span><span class="fu"><a href="../reference/plot_samples_present.html">plot_samples_present</a></span><span class="op">(</span><span class="va">samples_present</span>, <span class="va">rowvars_prot</span>, breaks<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">10</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Scale for <span style="color: #00BB00;">fill</span> is already present.</span></span>
<span><span class="co">#&gt; Adding another scale for <span style="color: #00BB00;">fill</span>, which will replace the existing scale.</span></span></code></pre></div>
<div class="figure">
<img src="LFQ_DDA_Peptide_QC_Summarisation_files/figure-html/unnamed-chunk-25-1.png" alt="Samples quantified for each protein at each level of processing" width="75%"><p class="caption">
Samples quantified for each protein at each level of processing
</p>
</div>
<p>From these two plots, we can see that the filtering to ensure that
all proteins have &gt;1 peptides removed only a few peptides, but more
proteins. Whether this is appropriate will depend on your data in
hand.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.5.1 (2025-06-13)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 24.04.2 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] dplyr_1.1.4                 tidyr_1.3.1                </span></span>
<span><span class="co">#&gt;  [3] ggplot2_3.5.2               biomasslmb_0.0.4           </span></span>
<span><span class="co">#&gt;  [5] QFeatures_1.18.0            MultiAssayExperiment_1.34.0</span></span>
<span><span class="co">#&gt;  [7] SummarizedExperiment_1.38.1 Biobase_2.68.0             </span></span>
<span><span class="co">#&gt;  [9] GenomicRanges_1.60.0        GenomeInfoDb_1.44.0        </span></span>
<span><span class="co">#&gt; [11] IRanges_2.42.0              S4Vectors_0.46.0           </span></span>
<span><span class="co">#&gt; [13] BiocGenerics_0.54.0         generics_0.1.4             </span></span>
<span><span class="co">#&gt; [15] MatrixGenerics_1.20.0       matrixStats_1.5.0          </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] DBI_1.2.3               gridExtra_2.3           rlang_1.1.6            </span></span>
<span><span class="co">#&gt;  [4] magrittr_2.0.3          clue_0.3-66             compiler_4.5.1         </span></span>
<span><span class="co">#&gt;  [7] RSQLite_2.4.1           png_0.1-8               systemfonts_1.2.3      </span></span>
<span><span class="co">#&gt; [10] vctrs_0.6.5             reshape2_1.4.4          stringr_1.5.1          </span></span>
<span><span class="co">#&gt; [13] ProtGenerics_1.40.0     pkgconfig_2.0.3         crayon_1.5.3           </span></span>
<span><span class="co">#&gt; [16] fastmap_1.2.0           backports_1.5.0         XVector_0.48.0         </span></span>
<span><span class="co">#&gt; [19] labeling_0.4.3          rmarkdown_2.29          UCSC.utils_1.4.0       </span></span>
<span><span class="co">#&gt; [22] visdat_0.6.0            ragg_1.4.0              UpSetR_1.4.0           </span></span>
<span><span class="co">#&gt; [25] purrr_1.0.4             bit_4.6.0               xfun_0.52              </span></span>
<span><span class="co">#&gt; [28] cachem_1.1.0            jsonlite_2.0.0          blob_1.2.4             </span></span>
<span><span class="co">#&gt; [31] DelayedArray_0.34.1     cluster_2.1.8.1         R6_2.6.1               </span></span>
<span><span class="co">#&gt; [34] bslib_0.9.0             stringi_1.8.7           RColorBrewer_1.1-3     </span></span>
<span><span class="co">#&gt; [37] genefilter_1.90.0       jquerylib_0.1.4         Rcpp_1.1.0             </span></span>
<span><span class="co">#&gt; [40] knitr_1.50              usethis_3.1.0           BiocBaseUtils_1.10.0   </span></span>
<span><span class="co">#&gt; [43] Matrix_1.7-3            splines_4.5.1           igraph_2.1.4           </span></span>
<span><span class="co">#&gt; [46] tidyselect_1.2.1        abind_1.4-8             yaml_2.3.10            </span></span>
<span><span class="co">#&gt; [49] lattice_0.22-7          tibble_3.3.0            plyr_1.8.9             </span></span>
<span><span class="co">#&gt; [52] withr_3.0.2             KEGGREST_1.48.1         evaluate_1.0.4         </span></span>
<span><span class="co">#&gt; [55] uniprotREST_1.0.0       desc_1.4.3              survival_3.8-3         </span></span>
<span><span class="co">#&gt; [58] Biostrings_2.76.0       pillar_1.11.0           corrplot_0.95          </span></span>
<span><span class="co">#&gt; [61] checkmate_2.3.2         rprojroot_2.0.4         scales_1.4.0           </span></span>
<span><span class="co">#&gt; [64] xtable_1.8-4            glue_1.8.0              lazyeval_0.2.2         </span></span>
<span><span class="co">#&gt; [67] tools_4.5.1             robustbase_0.99-4-1     annotate_1.86.1        </span></span>
<span><span class="co">#&gt; [70] fs_1.6.6                XML_3.99-0.18           grid_4.5.1             </span></span>
<span><span class="co">#&gt; [73] MsCoreUtils_1.20.0      AnnotationDbi_1.70.0    GenomeInfoDbData_1.2.14</span></span>
<span><span class="co">#&gt; [76] naniar_1.1.0            cli_3.6.5               textshaping_1.0.1      </span></span>
<span><span class="co">#&gt; [79] S4Arrays_1.8.1          AnnotationFilter_1.32.0 gtable_0.3.6           </span></span>
<span><span class="co">#&gt; [82] DEoptimR_1.1-3-1        sass_0.4.10             digest_0.6.37          </span></span>
<span><span class="co">#&gt; [85] SparseArray_1.8.0       htmlwidgets_1.6.4       farver_2.1.2           </span></span>
<span><span class="co">#&gt; [88] memoise_2.0.1           htmltools_0.5.8.1       pkgdown_2.1.3          </span></span>
<span><span class="co">#&gt; [91] lifecycle_1.0.4         httr_1.4.7              bit64_4.6.0-1          </span></span>
<span><span class="co">#&gt; [94] MASS_7.3-65</span></span></code></pre></div>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-http://zotero.org/users/5634351/items/FZN3QTTZ" class="csl-entry">
Sticker, Adriaan, Ludger Goeminne, Lennart Martens, and Lieven Clement.
2020. <span>“Robust Summarization and Inference in Proteome-wide
Label-free Quantification.”</span> <em>Molecular &amp; cellular
proteomics: MCP</em> 19 (7): 1209–19. <a href="https://doi.org/10.1074/mcp.RA119.001624" class="external-link">https://doi.org/10.1074/mcp.RA119.001624</a>.
</div>
</div>
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Tom Smith.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
