<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>TMT QC PSM-level quantification and summarisation to protein-level abundance • biomasslmb</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="TMT QC PSM-level quantification and summarisation to protein-level abundance">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">biomasslmb</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/LFQ_DDA_Peptide_QC_Summarisation.html">LFQ-DDA QC Peptide-level quantification and summarisation to protein-level abundance</a></li>
    <li><a class="dropdown-item" href="../articles/LFQ_DIA_Peptide_QC.html">LFQ-DIA QC Peptide-level quantification and summarisation to protein-level abundance</a></li>
    <li><a class="dropdown-item" href="../articles/TMT_PSM_QC_Summarisation.html">TMT QC PSM-level quantification and summarisation to protein-level abundance</a></li>
    <li><a class="dropdown-item" href="../articles/TMT_summarisation_methods.html">TMT - comparing protein summarisation approaches</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>TMT QC PSM-level quantification and summarisation to protein-level abundance</h1>
                        <h4 data-toc-skip class="author">Tom Smith</h4>
            
            <h4 data-toc-skip class="date">2025-03-28</h4>
      

      <div class="d-none name"><code>TMT_PSM_QC_Summarisation.Rmd</code></div>
    </div>

    
    
<p>Quantitative proteomics using isobaric tagging such as Tandem Mass
Tags (TMT) has a considerable benefit over Label-Free Quantification
(LFQ) in that up to 35 samples can be quantified for each Peptide
Spectrum Match (PSM). This has multiple benefits over analysing samples
in separate runs (LFQ):</p>
<ol style="list-style-type: decimal">
<li>TMT reduces protein quantification variance since PSM-level
quantification is derived from the same MS1 ion for all samples</li>
<li>LFQ suffers from much higher missing values when comparing across
samples due to the limited number of ions that can be fragmented in each
run and the associated issue of peptides being identified in only a
subset of runs <span class="citation">(O’Connell et al. 2018)</span>.
This is ameliorated to a significant degree by Data-Independent
Aquisition (DIA) LFQ. However, DIA still involves quantifying each
sample separately, so missing values are not entirely removed, and
proteins in each sample may be quantified from different sets of
peptides.</li>
</ol>
<p>Because TMT quantifies from the same MS1 ion for all samples, this
standardises the features quantified in each sample, which simplifies
the comparison between samples and increases quantification accuracy of
summarised features such as proteins.</p>
<p>However, TMT does suffer from ratio compression, which should be
avoiding by performing quantification with SPS MS3 <span class="citation">(McAlister et al. 2014)</span>.</p>
<p>Here, we will QC and filter the PSM level abundances from PD before
summarising them to protein-level abundances.</p>
<div class="section level2">
<h2 id="load-required-packages">Load required packages<a class="anchor" aria-label="anchor" href="#load-required-packages"></a>
</h2>
<p>To clarify which functionality is provided by which package, we will
use <code>package::function</code>. For your own code, there is no need
to specify the package unless you want to maintain this clarity.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/RforMassSpectrometry/QFeatures" class="external-link">QFeatures</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://lmb-mass-spec-compbio.github.io/biomasslmb/">biomasslmb</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org" class="external-link">tidyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="defining-the-contaminant-proteins">Defining the contaminant proteins<a class="anchor" aria-label="anchor" href="#defining-the-contaminant-proteins"></a>
</h2>
<p>We need to remove contaminant proteins. These were defined here using
the cRAP database. Below, we parse the contaminants fasta to extract the
IDs for the proteins in both ‘cRAP’ format and Uniprot IDs.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">crap_fasta_inf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span></span>
<span>  <span class="st">"extdata"</span>, <span class="st">"cRAP_20190401.fasta.gz"</span>, </span>
<span>  package <span class="op">=</span> <span class="st">"biomasslmb"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Extract the protein IDs associated with each cRAP protein</span></span>
<span><span class="va">crap_accessions</span> <span class="op">&lt;-</span> <span class="fu">biomasslmb</span><span class="fu">::</span><span class="fu"><a href="../reference/get_crap_fasta_accessions.html">get_crap_fasta_accessions</a></span><span class="op">(</span><span class="va">crap_fasta_inf</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">crap_accessions</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "cRAP001" "P00330"  "cRAP002" "P02768"  "cRAP003" "P00883"</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="read-in-input-data">Read in input data<a class="anchor" aria-label="anchor" href="#read-in-input-data"></a>
</h2>
<p>We start by reading in quantitative proteomics data into a
<code>QFeatures</code> object, which is the standard Bioconductor object
for holding quantitative proteomics data. See <a href="https://www.bioconductor.org/packages/release/bioc/html/QFeatures.html" class="external-link">here</a>
for documentation about the <code>QFeatures</code> object.In this case,
we are not adding any experimental details to the <code>QFeatures</code>
object, so the <code>ColData</code> will be empty.</p>
<p><code>psm_tmt_total</code> is a data set available from the
<code>biomasslmb</code> package containing the PSM-level output from
Proteome Discoverer (PD) for an experiment with 10 samples (each one
being a separate TMT tags). It is a truncated file containing the first
5,000 PSMs only. Here, we will not include any details about the
experimental design. Usually, these would be included by providing a
<code>data.frame</code> to the <code>colData</code> argument. See
<code><a href="https://rdrr.io/pkg/QFeatures/man/readQFeatures.html" class="external-link">?readQFeatures</a></code> for full details on how to read the
quantification data in a <code>QFeatures</code> object.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">tmt_qf</span> <span class="op">&lt;-</span> <span class="fu">QFeatures</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/readQFeatures.html" class="external-link">readQFeatures</a></span><span class="op">(</span>assayData <span class="op">=</span> <span class="va">psm_tmt_total</span>,</span>
<span>                                   quantCols <span class="op">=</span> <span class="fl">36</span><span class="op">:</span><span class="fl">45</span>, </span>
<span>                                   name <span class="op">=</span> <span class="st">"psms_raw"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Checking arguments.</span></span>
<span><span class="co">#&gt; Loading data as a 'SummarizedExperiment' object.</span></span>
<span><span class="co">#&gt; Formatting sample annotations (colData).</span></span>
<span><span class="co">#&gt; Formatting data as a 'QFeatures' object.</span></span></code></pre></div>
<p>We have 2 quantification values which are exactly zero. These should
be replaced with NA, since mass spectrometry is not capable of asserting
that the protein had zero abundance in the sample.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'psms_raw'</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">QFeatures</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-missing-data.html" class="external-link">zeroIsNA</a></span><span class="op">(</span><span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'psms_raw'</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p>We first perform routine filtering to remove PSMs that:</p>
<ul>
<li>Could originate from contaminants. See
<code><a href="../reference/filter_features_pd_dda.html">?filter_features_pd_dda</a></code> for further details, including the
removal of ‘associated cRAP’.</li>
<li>Don’t have a unique master protein.</li>
</ul>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Perform routine raw data filtering.</span></span>
<span><span class="co"># - Remove PSMs from contaminant proteins</span></span>
<span><span class="co"># - Remove PSMs where protein ID is empty or not unique</span></span>
<span><span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'psms_filtered'</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filter_features_pd_dda.html">filter_features_pd_dda</a></span><span class="op">(</span><span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'psms_raw'</span><span class="op">]</span><span class="op">]</span>,</span>
<span>                                             crap_proteins<span class="op">=</span><span class="va">crap_accessions</span>,</span>
<span>                                             level<span class="op">=</span><span class="st">'PSM'</span>,</span>
<span>                                             TMT<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>                                             filter_crap<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>                                             filter_associated_crap<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>                                             unique_master<span class="op">=</span><span class="cn">TRUE</span>, <span class="op">)</span></span>
<span><span class="co">#&gt; Filtering data...</span></span>
<span><span class="co">#&gt; 5000 features found from 2364 master proteins =&gt; Input</span></span>
<span><span class="co">#&gt; 242 contaminant proteins supplied</span></span>
<span><span class="co">#&gt; 55 proteins identified as 'contaminant associated'</span></span>
<span><span class="co">#&gt; 4937 features found from 2347 master proteins =&gt; contaminant features removed</span></span>
<span><span class="co">#&gt; 4931 features found from 2342 master proteins =&gt; associated contaminant features removed</span></span>
<span><span class="co">#&gt; 4929 features found from 2341 master proteins =&gt; features without a master protein removed</span></span>
<span><span class="co">#&gt; 4754 features found from 2243 master proteins =&gt; features with non-unique master proteins removed</span></span>
<span><span class="co">#&gt; 4753 features found from 2243 master proteins =&gt; features without quantification removed</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="normalise">Normalise<a class="anchor" aria-label="anchor" href="#normalise"></a>
</h2>
<p>Next, we plot the peptide intensities to check they are approximately
the same using <code><a href="../reference/plot_quant.html">biomasslmb::plot_quant</a></code>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Plot the peptide-level quantification distributions per sample</span></span>
<span><span class="fu">biomasslmb</span><span class="fu">::</span><span class="fu"><a href="../reference/plot_quant.html">plot_quant</a></span><span class="op">(</span><span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'psms_filtered'</span><span class="op">]</span><span class="op">]</span>,</span>
<span>                       log2transform<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>                       method<span class="op">=</span><span class="st">'density'</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="../reference/theme_biomasslmb.html">theme_biomasslmb</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">'PSM abundance (log2)'</span><span class="op">)</span></span></code></pre></div>
<p><img src="TMT_PSM_QC_Summarisation_files/figure-html/unnamed-chunk-6-1.png" width="75%"></p>
<p>Since the same amount of sample was labelled in each case, it’s
reasonable to use ‘diff.median’ normalisation with
<code><a href="https://rdrr.io/pkg/BiocGenerics/man/normalize.html" class="external-link">QFeatures::normalize</a></code>. First though, we need to
log-transform the abundances so the normalisation works as expected. We
exponentiate the quantification values back to the original scale
(reverse the log-transformation) afterwards, since the downstream steps
including summarisation to protein require non-transformed values</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Normalise the log2-transformed abundances using diff.median</span></span>
<span><span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'psms_filtered_norm'</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">QFeatures</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/normalize.html" class="external-link">normalize</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-processing.html" class="external-link">logTransform</a></span><span class="op">(</span><span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'psms_filtered'</span><span class="op">]</span><span class="op">]</span>, base<span class="op">=</span><span class="fl">2</span><span class="op">)</span>, method<span class="op">=</span><span class="st">'diff.median'</span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># Exponentiate the quantification values back to the initial scale.</span></span>
<span><span class="fu">assay</span><span class="op">(</span><span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'psms_filtered_norm'</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fl">2</span><span class="op">^</span><span class="fu">assay</span><span class="op">(</span><span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'psms_filtered_norm'</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot the peptide-level quantification distributions per sample</span></span>
<span><span class="fu">biomasslmb</span><span class="fu">::</span><span class="fu"><a href="../reference/plot_quant.html">plot_quant</a></span><span class="op">(</span><span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'psms_filtered_norm'</span><span class="op">]</span><span class="op">]</span>,</span>
<span>                       log2transform<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>                       method<span class="op">=</span><span class="st">'density'</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="../reference/theme_biomasslmb.html">theme_biomasslmb</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">'PSM abundance (log2)'</span><span class="op">)</span></span></code></pre></div>
<p><img src="TMT_PSM_QC_Summarisation_files/figure-html/unnamed-chunk-7-1.png" width="75%"></p>
</div>
<div class="section level2">
<h2 id="removing-low-quality-psms">Removing low quality PSMs<a class="anchor" aria-label="anchor" href="#removing-low-quality-psms"></a>
</h2>
<p>We want to remove low Signal:Noise (S:N) PSMs, since the
quantification values will be less accurate and there will be more
missing values. We can inspect the relationship between S:N and missing
values using the <code>plot_missing_SN</code> function.</p>
<p>Note that where the signal:noise &gt; 10, there are far fewer missing
values.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Add a more accurate average S:N ratio value.</span></span>
<span><span class="co"># The one calculated by PD doesn't treat NA values appropriately!</span></span>
<span><span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'psms_filtered_norm'</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/update_average_sn.html">update_average_sn</a></span><span class="op">(</span><span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'psms_filtered_norm'</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plot_missing_SN.html">plot_missing_SN</a></span><span class="op">(</span><span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'psms_filtered_norm'</span><span class="op">]</span><span class="op">]</span>, bins <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="TMT_PSM_QC_Summarisation_files/figure-html/unnamed-chunk-8-1.png" alt="Missing values per PSM, in relation to the signal:noise ratio" width="75%"><p class="caption">
Missing values per PSM, in relation to the signal:noise ratio
</p>
</div>
<p>We can also look into this relationship at the tag level using
<code>plot_missing_SN_per_sample</code> to inspect whether there was a
labeling issue with any particular tag. In this case, there is no tag
which appears to have a high proportion of missing values when
signal:noise &gt; 10.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_missing_SN_per_sample.html">plot_missing_SN_per_sample</a></span><span class="op">(</span><span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'psms_filtered_norm'</span><span class="op">]</span><span class="op">]</span>, bins <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="TMT_PSM_QC_Summarisation_files/figure-html/unnamed-chunk-9-1.png" alt="Missing values per tag, in relation to the signal:noise ratio" width="75%"><p class="caption">
Missing values per tag, in relation to the signal:noise ratio
</p>
</div>
<p>Based on the above, we will filter the PSMs to only retain those with
S:N &gt; 10 using <code>filter_TMT_PSMs</code> since these PSMs are
clearly much closer to the limit of detection for the MS and therefore
likely to contain less accurate quantification data. Note that this only
necessitates removing ~ 1/20 of the PSMs in this case. Using the same
function, we will also remove PSMs with interference/co-isolation
&gt;50% since these are also likely to contain less accurate
quantification data.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Then filter PSMs to remove low S:N and/or high interference</span></span>
<span><span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'psms_filtered_sn'</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filter_TMT_PSMs.html">filter_TMT_PSMs</a></span><span class="op">(</span><span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'psms_filtered_norm'</span><span class="op">]</span><span class="op">]</span>,</span>
<span>                                                inter_thresh<span class="op">=</span><span class="fl">50</span>, sn_thresh<span class="op">=</span><span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt; Filtering PSMs...</span></span>
<span><span class="co">#&gt; 4753 features found from 2243 master proteins =&gt; Initial PSMs</span></span>
<span><span class="co">#&gt; 4753 features found from 2243 master proteins =&gt; Removing PSMs without quantification values</span></span>
<span><span class="co">#&gt; 4633 features found from 2207 master proteins =&gt; Removing PSMs with high Co-isolation/interference</span></span>
<span><span class="co">#&gt; 4455 features found from 2173 master proteins =&gt; Removing PSMs with low average S:N ratio</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="summarising-to-protein-level-abundances">Summarising to protein-level abundances<a class="anchor" aria-label="anchor" href="#summarising-to-protein-level-abundances"></a>
</h2>
<p>Now that we have inspected the PSM-level quantification and filtered
the PSMs, we can summarise the PSMs to protein-level abundances.</p>
<p>For PSM to protein summarisation with TMT quantification, simply
summing together the PSM-level abundances provides accurate estimates,
so long as there are no missing values. If many missing values are
present, the <code><a href="https://rdrr.io/pkg/MsCoreUtils/man/robustSummary.html" class="external-link">MsCoreUtils::robustSummary</a></code> method may be
preferred, since it is able to summarise accurately, even in the
presence of missing values <span class="citation">(Sticker et al.
2020)</span>. For a comparison of the two summarisation approaches, see
the <code>TMT - comparing protein summarisation approaches</code>
vignette.</p>
<p>Here, we will use the simple sum summarisation. First, we need to
consider the missing values. If we leave the missing values in, we
either need to allow some protein quantifications to be <code>NA</code>,
or else ignore the missing values in the summarisation. The latter would
potentially mean different PSMs are summed for the a given protein
across the samples, making a given protein’s summarised abundances
across the samples less accurately estimated. Here, we will therefore
remove all PSMs with missing values.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'psms_filtered_missing'</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">QFeatures</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">filterNA</a></span><span class="op">(</span></span>
<span>  <span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'psms_filtered_sn'</span><span class="op">]</span><span class="op">]</span>, <span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<p>Next, we will remove PSMs for proteins with fewer than 2 PSMs. This
is a common filter in proteomics to ensure the protein-level
quantifications are derived from at least two independent observations.
In some cases, for example phosphoproteomics, or where this filter
appears to be too stringent, it may be appropriate to skip it.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">min_psms</span> <span class="op">&lt;-</span> <span class="fl">2</span></span>
<span><span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'psms_filtered_forSum'</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">biomasslmb</span><span class="fu">::</span><span class="fu"><a href="../reference/filter_features_per_protein.html">filter_features_per_protein</a></span><span class="op">(</span></span>
<span>  <span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'psms_filtered_missing'</span><span class="op">]</span><span class="op">]</span>, min_features <span class="op">=</span> <span class="va">min_psms</span><span class="op">)</span></span></code></pre></div>
<p>Finally, we perform the summarisation.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Aggregate to protein-level abundances (using QFeatures function)</span></span>
<span><span class="va">tmt_qf</span> <span class="op">&lt;-</span> <span class="fu">QFeatures</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">aggregateFeatures</a></span><span class="op">(</span><span class="va">tmt_qf</span>, </span>
<span>                                       i <span class="op">=</span> <span class="st">"psms_filtered_forSum"</span>, </span>
<span>                                       fcol <span class="op">=</span> <span class="st">"Master.Protein.Accessions"</span>,</span>
<span>                                       name <span class="op">=</span> <span class="st">"protein"</span>,</span>
<span>                                       fun <span class="op">=</span> <span class="fu">base</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Your row data contain missing values. Please read the relevant</span></span>
<span><span class="co">#&gt; section(s) in the aggregateFeatures manual page regarding the effects</span></span>
<span><span class="co">#&gt; of missing values on data aggregation.</span></span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'protein'</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">QFeatures</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-processing.html" class="external-link">logTransform</a></span><span class="op">(</span></span>
<span>  <span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'protein'</span><span class="op">]</span><span class="op">]</span>, base<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="inspecting-the-number-of-psms-and-proteins-through-the-processing-steps">Inspecting the number of PSMs and proteins through the processing
steps<a class="anchor" aria-label="anchor" href="#inspecting-the-number-of-psms-and-proteins-through-the-processing-steps"></a>
</h3>
<p>Now that we have performed all the filtering steps and summarisation
to protein-level abundances, it’s helpful to visualise how many
PSMs/proteins were retained at each level of the processing. We can use
the <code><a href="../reference/get_samples_present.html">biomasslmb::get_samples_present</a></code> and
<code><a href="../reference/plot_samples_present.html">biomasslmb::plot_samples_present</a></code> functions for this. First
though, we need to decide which ‘experiments’ we want to plot and define
a named character vector since the QFeatures names are not sufficiently
clear by themselves</p>
<p>Below, we inspect the experiment names.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">tmt_qf</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "psms_raw"              "psms_filtered"         "psms_filtered_norm"   </span></span>
<span><span class="co">#&gt; [4] "psms_filtered_sn"      "psms_filtered_missing" "psms_filtered_forSum" </span></span>
<span><span class="co">#&gt; [7] "protein"</span></span></code></pre></div>
<p>In this case, we don’t want to plot the
<code>psms_filterered_norm</code>, since that’s just the normalised
quantification present in <code>psms_filtered</code> and no PSMs were
removed in that step.</p>
<div class="section level4">
<h4 id="samples-per-psm">Samples per PSM<a class="anchor" aria-label="anchor" href="#samples-per-psm"></a>
</h4>
<p>We’ll start by inspecting the number of PSMs in each experiment. We
therefore define a named character vector all the PSM-level experiments,
excluding <code>psms_filtered_norm</code>. We set the row variables to
be <code>Sequence</code>, <code>Modifications</code> and
<code>RT.in.min</code> so that we count the number of unique PSMs.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span></span>
<span><span class="va">rename_cols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'All PSMs'</span> <span class="op">=</span> <span class="st">'psms_raw'</span> ,</span>
<span>                 <span class="st">'Quantified, contaminants removed'</span> <span class="op">=</span> <span class="st">'psms_filtered'</span>,</span>
<span>                 <span class="st">'Signal:Noise &gt; 10'</span> <span class="op">=</span> <span class="st">'psms_filtered_sn'</span>,</span>
<span>                 <span class="st">'No missing values'</span> <span class="op">=</span> <span class="st">'psms_filtered_missing'</span>,</span>
<span>                 <span class="st">'&gt;1 PSM per protein'</span> <span class="op">=</span> <span class="st">'psms_filtered_forSum'</span><span class="op">)</span></span>
<span></span>
<span><span class="va">rowvars</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'Sequence'</span>, <span class="st">'Modifications'</span>, <span class="st">'RT.in.min'</span><span class="op">)</span></span>
<span></span>
<span><span class="va">samples_present</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_samples_present.html">get_samples_present</a></span><span class="op">(</span><span class="va">tmt_qf</span><span class="op">[</span>,,<span class="fu"><a href="https://rdrr.io/r/base/unname.html" class="external-link">unname</a></span><span class="op">(</span><span class="va">rename_cols</span><span class="op">)</span><span class="op">]</span>, <span class="va">rowvars</span>, <span class="va">rename_cols</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: 'experiments' dropped; see 'drops()'</span></span>
<span><span class="co">#&gt; harmonizing input:</span></span>
<span><span class="co">#&gt;   removing 20 sampleMap rows not in names(experiments)</span></span>
<span><span class="fu"><a href="../reference/plot_samples_present.html">plot_samples_present</a></span><span class="op">(</span><span class="va">samples_present</span>, <span class="va">rowvars</span>, breaks<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">10</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">'PSM'</span><span class="op">)</span></span>
<span><span class="co">#&gt; Scale for <span style="color: #00BB00;">fill</span> is already present.</span></span>
<span><span class="co">#&gt; Adding another scale for <span style="color: #00BB00;">fill</span>, which will replace the existing scale.</span></span></code></pre></div>
<div class="figure">
<img src="TMT_PSM_QC_Summarisation_files/figure-html/unnamed-chunk-16-1.png" alt="Samples quantified for each PSM at each level of processing" width="75%"><p class="caption">
Samples quantified for each PSM at each level of processing
</p>
</div>
</div>
<div class="section level4">
<h4 id="samples-per-protein">Samples per Protein<a class="anchor" aria-label="anchor" href="#samples-per-protein"></a>
</h4>
<p>Next, we’ll use the same functions to inspect the number of proteins
at each level of processing. We need to supply an updated named
character vector to include the <code>protein</code> experiment and set
the row variables to be just the <code>Master.Protein.Accesions</code>
column.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span></span>
<span><span class="va">rename_cols_prot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">rename_cols</span>, <span class="st">'Protein'</span><span class="op">=</span><span class="st">'protein'</span><span class="op">)</span></span>
<span></span>
<span><span class="va">rowvars_prot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'Master.Protein.Accessions'</span><span class="op">)</span></span>
<span></span>
<span><span class="va">samples_present</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_samples_present.html">get_samples_present</a></span><span class="op">(</span><span class="va">tmt_qf</span>, <span class="va">rowvars_prot</span>, <span class="va">rename_cols_prot</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plot_samples_present.html">plot_samples_present</a></span><span class="op">(</span><span class="va">samples_present</span>, <span class="va">rowvars_prot</span>, breaks<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">10</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Scale for <span style="color: #00BB00;">fill</span> is already present.</span></span>
<span><span class="co">#&gt; Adding another scale for <span style="color: #00BB00;">fill</span>, which will replace the existing scale.</span></span></code></pre></div>
<div class="figure">
<img src="TMT_PSM_QC_Summarisation_files/figure-html/unnamed-chunk-17-1.png" alt="Samples quantified for each protein at each level of processing" width="75%"><p class="caption">
Samples quantified for each protein at each level of processing
</p>
</div>
<p>In this case, because we have removed all missing values, the last
two experiments have identical counts for the number of samples for each
protein. From these two plots, we can see that the filtering to ensure
that all proteins have &gt;1 PSM removed only a few PSMs, but many
proteins. Whether this is appropriate will depend on your data in
hand.</p>
</div>
</div>
<div class="section level3">
<h3 id="summary">Summary<a class="anchor" aria-label="anchor" href="#summary"></a>
</h3>
<p>Here, we have QCed the PSMs and filtered them in the following
steps:</p>
<ul>
<li>Remove PSMs that are likely from contaminants</li>
<li>Remove PSMs without a unique ‘master protein’</li>
<li>Remove PSMs without any quantification values</li>
<li>Normalise the abundance values so they have the same median value in
all samples</li>
<li>Remove PSMs with very low signal:noise</li>
<li>Remove PSMs with high interference</li>
</ul>
<p>We then summarised the PSM level abundances with two approaches,
<code>sum</code> and <code>robustSummary</code>, before comparing the
protein-level abundances obtained with each of them.
<code>robustSummary</code> handles missing values and therefore enables
more proteins to be quantified. <code>sum</code> is sensitive to
outliers with very high abundance, while <code>robust</code> is more
sensitive to outliers with very low abundance.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.4.3 (2025-02-28)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 24.04.2 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] dplyr_1.1.4                 tidyr_1.3.1                </span></span>
<span><span class="co">#&gt;  [3] ggplot2_3.5.1               biomasslmb_0.0.1           </span></span>
<span><span class="co">#&gt;  [5] QFeatures_1.16.0            MultiAssayExperiment_1.32.0</span></span>
<span><span class="co">#&gt;  [7] SummarizedExperiment_1.36.0 Biobase_2.66.0             </span></span>
<span><span class="co">#&gt;  [9] GenomicRanges_1.58.0        GenomeInfoDb_1.42.3        </span></span>
<span><span class="co">#&gt; [11] IRanges_2.40.1              S4Vectors_0.44.0           </span></span>
<span><span class="co">#&gt; [13] BiocGenerics_0.52.0         MatrixGenerics_1.18.1      </span></span>
<span><span class="co">#&gt; [15] matrixStats_1.5.0          </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] DBI_1.2.3               rlang_1.1.5             magrittr_2.0.3         </span></span>
<span><span class="co">#&gt;  [4] clue_0.3-66             compiler_4.4.3          RSQLite_2.3.9          </span></span>
<span><span class="co">#&gt;  [7] png_0.1-8               systemfonts_1.2.1       vctrs_0.6.5            </span></span>
<span><span class="co">#&gt; [10] reshape2_1.4.4          stringr_1.5.1           ProtGenerics_1.38.0    </span></span>
<span><span class="co">#&gt; [13] pkgconfig_2.0.3         crayon_1.5.3            fastmap_1.2.0          </span></span>
<span><span class="co">#&gt; [16] XVector_0.46.0          labeling_0.4.3          rmarkdown_2.29         </span></span>
<span><span class="co">#&gt; [19] UCSC.utils_1.2.0        visdat_0.6.0            ragg_1.3.3             </span></span>
<span><span class="co">#&gt; [22] purrr_1.0.4             bit_4.6.0               xfun_0.51              </span></span>
<span><span class="co">#&gt; [25] zlibbioc_1.52.0         cachem_1.1.0            jsonlite_2.0.0         </span></span>
<span><span class="co">#&gt; [28] blob_1.2.4              DelayedArray_0.32.0     cluster_2.1.8          </span></span>
<span><span class="co">#&gt; [31] R6_2.6.1                bslib_0.9.0             stringi_1.8.7          </span></span>
<span><span class="co">#&gt; [34] genefilter_1.88.0       jquerylib_0.1.4         Rcpp_1.0.14            </span></span>
<span><span class="co">#&gt; [37] knitr_1.50              usethis_3.1.0           BiocBaseUtils_1.8.0    </span></span>
<span><span class="co">#&gt; [40] Matrix_1.7-2            splines_4.4.3           igraph_2.1.4           </span></span>
<span><span class="co">#&gt; [43] tidyselect_1.2.1        abind_1.4-8             yaml_2.3.10            </span></span>
<span><span class="co">#&gt; [46] lattice_0.22-6          tibble_3.2.1            plyr_1.8.9             </span></span>
<span><span class="co">#&gt; [49] withr_3.0.2             KEGGREST_1.46.0         evaluate_1.0.3         </span></span>
<span><span class="co">#&gt; [52] desc_1.4.3              survival_3.8-3          Biostrings_2.74.1      </span></span>
<span><span class="co">#&gt; [55] pillar_1.10.1           corrplot_0.95           generics_0.1.3         </span></span>
<span><span class="co">#&gt; [58] rprojroot_2.0.4         munsell_0.5.1           scales_1.3.0           </span></span>
<span><span class="co">#&gt; [61] xtable_1.8-4            glue_1.8.0              lazyeval_0.2.2         </span></span>
<span><span class="co">#&gt; [64] tools_4.4.3             robustbase_0.99-4-1     annotate_1.84.0        </span></span>
<span><span class="co">#&gt; [67] fs_1.6.5                XML_3.99-0.18           grid_4.4.3             </span></span>
<span><span class="co">#&gt; [70] cutr_0.0.0.9000         MsCoreUtils_1.18.0      AnnotationDbi_1.68.0   </span></span>
<span><span class="co">#&gt; [73] colorspace_2.1-1        GenomeInfoDbData_1.2.13 naniar_1.1.0           </span></span>
<span><span class="co">#&gt; [76] cli_3.6.4               textshaping_1.0.0       S4Arrays_1.6.0         </span></span>
<span><span class="co">#&gt; [79] AnnotationFilter_1.30.0 gtable_0.3.6            DEoptimR_1.1-3-1       </span></span>
<span><span class="co">#&gt; [82] sass_0.4.9              digest_0.6.37           SparseArray_1.6.2      </span></span>
<span><span class="co">#&gt; [85] htmlwidgets_1.6.4       farver_2.1.2            memoise_2.0.1          </span></span>
<span><span class="co">#&gt; [88] htmltools_0.5.8.1       pkgdown_2.1.1           lifecycle_1.0.4        </span></span>
<span><span class="co">#&gt; [91] httr_1.4.7              bit64_4.6.0-1           MASS_7.3-64</span></span></code></pre></div>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-http://zotero.org/users/5634351/items/ZK69WYZ2" class="csl-entry">
McAlister, Graeme C., David P. Nusinow, Mark P. Jedrychowski, Martin
Wühr, Edward L. Huttlin, Brian K. Erickson, Ramin Rad, Wilhelm Haas, and
Steven P. Gygi. 2014. <span>“MultiNotch MS3 Enables Accurate, Sensitive,
and Multiplexed Detection of Differential Expression Across Cancer Cell
Line Proteomes.”</span> <em>Analytical Chemistry</em> 86 (14): 7150–58.
<a href="https://doi.org/10.1021/ac502040v" class="external-link">https://doi.org/10.1021/ac502040v</a>.
</div>
<div id="ref-http://zotero.org/users/5634351/items/ME322P2A" class="csl-entry">
O’Connell, Jeremy D., Joao A. Paulo, Jonathon J. O’Brien, and Steven P.
Gygi. 2018. <span>“Proteome-Wide Evaluation of Two Common Protein
Quantification Methods.”</span> <em>Journal of Proteome Research</em> 17
(5): 1934–42. <a href="https://doi.org/10.1021/acs.jproteome.8b00016" class="external-link">https://doi.org/10.1021/acs.jproteome.8b00016</a>.
</div>
<div id="ref-http://zotero.org/users/5634351/items/FZN3QTTZ" class="csl-entry">
Sticker, Adriaan, Ludger Goeminne, Lennart Martens, and Lieven Clement.
2020. <span>“Robust Summarization and Inference in Proteome-wide
Label-free Quantification.”</span> <em>Molecular &amp; cellular
proteomics: MCP</em> 19 (7): 1209–19. <a href="https://doi.org/10.1074/mcp.RA119.001624" class="external-link">https://doi.org/10.1074/mcp.RA119.001624</a>.
</div>
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Tom Smith.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
