<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>TMT QC PSM-level quantification and summarisation to protein-level abundance • biomasslmb</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="TMT QC PSM-level quantification and summarisation to protein-level abundance">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">biomasslmb</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/TMT_PSM_QC_Summarisation.html">TMT QC PSM-level quantification and summarisation to protein-level abundance</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>TMT QC PSM-level quantification and summarisation to protein-level abundance</h1>
                        <h4 data-toc-skip class="author">Tom Smith</h4>
            
            <h4 data-toc-skip class="date">2025-02-18</h4>
      

      <div class="d-none name"><code>TMT_PSM_QC_Summarisation.Rmd</code></div>
    </div>

    
    
<p>Quantitative proteomics using isobaric tagging such as Tandem Mass
Tags (TMT) has a considerable benefit over Label-Free Quantification
(LFQ) in that up to 35 samples can be quantified for each Peptide
Spectrum Match (PSM). This has multiple benefits over analysing samples
in separate runs (LFQ):</p>
<ol style="list-style-type: decimal">
<li>TMT reduces protein quantification variance since PSM-level
quantification is derived from the same MS1 ion for all samples</li>
<li>LFQ suffers from much higher missing values when comparing across
samples due to the limited number of ions that can be fragmented in each
run and the associated issue of peptides being identified in only a
subset of runs <span class="citation">(O’Connell et al. 2018)</span>.
This is ameliorated to a significant degree by Data-Independent
Aquisition (DIA) LFQ. However, DIA still involves quantifying each
sample separately, so missing values are not entirely removed, and
proteins in each sample may be quantified from different sets of
peptides.</li>
</ol>
<p>Because TMT quantifies from the same MS1 ion for all samples, this
standardises the features quantified in each sample, which simplifies
the comparison between samples and increases quantification accuracy of
summarised features such as proteins.</p>
<p>However, TMT does suffer from ratio compression, which should be
avoiding by performing quantification with SPS MS3 <span class="citation">(McAlister et al. 2014)</span>.</p>
<p>Here, we will QC and filter the PSM level abundances from PD before
summarising them to protein-level abundances.</p>
<div class="section level2">
<h2 id="load-required-packages">Load required packages<a class="anchor" aria-label="anchor" href="#load-required-packages"></a>
</h2>
<p>To clarify which functionality is provided by which package, we will
use <code>package::function</code>. For your own code, there is no need
to specify the package unless you want to maintain this clarity.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/RforMassSpectrometry/QFeatures" class="external-link">QFeatures</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://lmb-mass-spec-compbio.github.io/biomasslmb/">biomasslmb</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org" class="external-link">tidyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="defining-the-contaminant-proteins">Defining the contaminant proteins<a class="anchor" aria-label="anchor" href="#defining-the-contaminant-proteins"></a>
</h2>
<p>We need to remove contaminant proteins. These were defined here using
the cRAP database. Below, we parse the contaminants fasta to extract the
IDs for the proteins in both ‘cRAP’ format and Uniprot IDs.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">crap_fasta_inf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span></span>
<span>  <span class="st">"extdata"</span>, <span class="st">"cRAP_20190401.fasta.gz"</span>, </span>
<span>  package <span class="op">=</span> <span class="st">"biomasslmb"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Extract the protein IDs associated with each cRAP protein</span></span>
<span><span class="va">crap_accessions</span> <span class="op">&lt;-</span> <span class="fu">biomasslmb</span><span class="fu">::</span><span class="fu"><a href="../reference/get_crap_fasta_accessions.html">get_crap_fasta_accessions</a></span><span class="op">(</span><span class="va">crap_fasta_inf</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">crap_accessions</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "cRAP001" "P00330"  "cRAP002" "P02768"  "cRAP003" "P00883"</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="read-in-input-data">Read in input data<a class="anchor" aria-label="anchor" href="#read-in-input-data"></a>
</h2>
<p>We start by reading in quantitative proteomics data into a
<code>QFeatures</code> object, which is the standard Bioconductor object
for holding quantitative proteomics data. See <a href="https://www.bioconductor.org/packages/release/bioc/html/QFeatures.html" class="external-link">here</a>
for documentation about the <code>QFeatures</code> object.In this case,
we are not adding any experimental details to the <code>QFeatures</code>
object, so the <code>ColData</code> will be empty.</p>
<p><code>psm_tmt_total</code> is a data set available from the
<code>biomasslmb</code> package containing the PSM-level output from
Proteome Discoverer (PD) for an experiment with 10 samples (each one
being a separate TMT tags). It is a truncated file containing the first
5,000 PSMs only. Here, we will not include any details about the
experimental design. Usually, these would be included by providing a
<code>data.frame</code> to the <code>colData</code> argument. See
<code><a href="https://rdrr.io/pkg/QFeatures/man/readQFeatures.html" class="external-link">?readQFeatures</a></code> for full details on how to read the
quantification data in a <code>QFeatures</code> object.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">tmt_qf</span> <span class="op">&lt;-</span> <span class="fu">QFeatures</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/readQFeatures.html" class="external-link">readQFeatures</a></span><span class="op">(</span>assayData <span class="op">=</span> <span class="va">psm_tmt_total</span>,</span>
<span>                                   quantCols <span class="op">=</span> <span class="fl">36</span><span class="op">:</span><span class="fl">45</span>, </span>
<span>                                   name <span class="op">=</span> <span class="st">"psms_raw"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Checking arguments.</span></span>
<span><span class="co">#&gt; Loading data as a 'SummarizedExperiment' object.</span></span>
<span><span class="co">#&gt; Formatting sample annotations (colData).</span></span>
<span><span class="co">#&gt; Formatting data as a 'QFeatures' object.</span></span>
<span></span>
<span><span class="co"># Add a more accurate average S:N ratio value.</span></span>
<span><span class="co"># The one calculated by PD doesn't treat NA values appropriately!</span></span>
<span><span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'psms_raw'</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/update_average_sn.html">update_average_sn</a></span><span class="op">(</span><span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'psms_raw'</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p>We have 2 quantification values which are exactly zero. These should
be replaced with NA, since mass spectrometry is not capable of asserting
that the protein had zero abundance in the sample.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'psms_raw'</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">QFeatures</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-missing-data.html" class="external-link">zeroIsNA</a></span><span class="op">(</span><span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'psms_raw'</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p>We first perform routine filtering to remove PSMs that:</p>
<ul>
<li>Could originate from contaminants. See
<code><a href="../reference/filter_features_pd_dda.html">?filter_features_pd_dda</a></code> for further details, including the
removal of ‘associated cRAP’.</li>
<li>Don’t have a unique master protein.</li>
</ul>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Perform routine raw data filtering.</span></span>
<span><span class="co"># - Remove PSMs from contaminant proteins</span></span>
<span><span class="co"># - Remove PSMs where protein ID is empty or not unique</span></span>
<span><span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'psms_filtered'</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filter_features_pd_dda.html">filter_features_pd_dda</a></span><span class="op">(</span><span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'psms_raw'</span><span class="op">]</span><span class="op">]</span>,</span>
<span>                                             crap_proteins<span class="op">=</span><span class="va">crap_accessions</span>,</span>
<span>                                             level<span class="op">=</span><span class="st">'PSM'</span>,</span>
<span>                                             TMT<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>                                             filter_crap<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>                                             filter_associated_crap<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>                                             unique_master<span class="op">=</span><span class="cn">TRUE</span>, <span class="op">)</span></span>
<span><span class="co">#&gt; Filtering data...</span></span>
<span><span class="co">#&gt; 5000 features found from 2364 master proteins =&gt; Input</span></span>
<span><span class="co">#&gt; 242 cRAP proteins supplied</span></span>
<span><span class="co">#&gt; 60 proteins identified as 'cRAP associated'</span></span>
<span><span class="co">#&gt; 4933 features found from 2346 master proteins =&gt; cRAP features removed</span></span>
<span><span class="co">#&gt; 4924 features found from 2340 master proteins =&gt; associated cRAP features removed</span></span>
<span><span class="co">#&gt; 4922 features found from 2339 master proteins =&gt; features without a master protein removed</span></span>
<span><span class="co">#&gt; 4748 features found from 2242 master proteins =&gt; features with non-unique master proteins removed</span></span>
<span><span class="co">#&gt; 4747 features found from 2242 master proteins =&gt; features without quantification removed</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="normalise">Normalise<a class="anchor" aria-label="anchor" href="#normalise"></a>
</h2>
<p>Next, we plot the peptide intensities to check they are approximately
the same using <code><a href="../reference/plot_quant.html">biomasslmb::plot_quant</a></code>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Plot the peptide-level quantification distributions per sample</span></span>
<span><span class="fu">biomasslmb</span><span class="fu">::</span><span class="fu"><a href="../reference/plot_quant.html">plot_quant</a></span><span class="op">(</span><span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'psms_filtered'</span><span class="op">]</span><span class="op">]</span>,</span>
<span>                       log2transform<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>                       method<span class="op">=</span><span class="st">'density'</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="../reference/theme_biomasslmb.html">theme_biomasslmb</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">'PSM abundance (log2)'</span><span class="op">)</span></span></code></pre></div>
<p><img src="TMT_PSM_QC_Summarisation_files/figure-html/unnamed-chunk-6-1.png" width="75%"></p>
<p>Since the same amount of sample was labelled in each case, it’s
reasonable to use ‘diff.median’ normalisation with
<code><a href="https://rdrr.io/pkg/BiocGenerics/man/normalize.html" class="external-link">QFeatures::normalize</a></code>. First though, we need to
log-transform the abundances so the normalisation works as expected. We
exponentiate the quantification values back to the original scale
(reverse the log-transformation) afterwards, since the downstream steps
including summarisation to protein require non-transformed values</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Normalise the log2-transformed abundances using diff.median</span></span>
<span><span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'psms_filtered_norm'</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">QFeatures</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/normalize.html" class="external-link">normalize</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-processing.html" class="external-link">logTransform</a></span><span class="op">(</span><span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'psms_filtered'</span><span class="op">]</span><span class="op">]</span>, base<span class="op">=</span><span class="fl">2</span><span class="op">)</span>, method<span class="op">=</span><span class="st">'diff.median'</span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># Exponentiate the quantification values back to the initial scale.</span></span>
<span><span class="fu">assay</span><span class="op">(</span><span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'psms_filtered_norm'</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fl">2</span><span class="op">^</span><span class="fu">assay</span><span class="op">(</span><span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'psms_filtered_norm'</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot the peptide-level quantification distributions per sample</span></span>
<span><span class="fu">biomasslmb</span><span class="fu">::</span><span class="fu"><a href="../reference/plot_quant.html">plot_quant</a></span><span class="op">(</span><span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'psms_filtered_norm'</span><span class="op">]</span><span class="op">]</span>,</span>
<span>                       log2transform<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>                       method<span class="op">=</span><span class="st">'density'</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="../reference/theme_biomasslmb.html">theme_biomasslmb</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">'PSM abundance (log2)'</span><span class="op">)</span></span></code></pre></div>
<p><img src="TMT_PSM_QC_Summarisation_files/figure-html/unnamed-chunk-7-1.png" width="75%"></p>
</div>
<div class="section level2">
<h2 id="removing-low-quality-psms">Removing low quality PSMs<a class="anchor" aria-label="anchor" href="#removing-low-quality-psms"></a>
</h2>
<p>We want to remove low Signal:Noise (S:N) PSMs, since the
quantification values will be less accurate and there will be more
missing values. We can inspect the relationship between S:N and missing
values using the <code>plot_missing_SN</code> function.</p>
<p>Note that where the signal:noise &gt; 10, there are far fewer missing
values.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu"><a href="../reference/plot_missing_SN.html">plot_missing_SN</a></span><span class="op">(</span><span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'psms_filtered_norm'</span><span class="op">]</span><span class="op">]</span>, bins <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="TMT_PSM_QC_Summarisation_files/figure-html/unnamed-chunk-8-1.png" alt="Missing values per PSM, in relation to the signal:noise ratio" width="75%"><p class="caption">
Missing values per PSM, in relation to the signal:noise ratio
</p>
</div>
<p>We can also look into this relationship at the tag level using
<code>plot_missing_SN_per_sample</code> to inspect whether there was a
labeling issue with any particular tag. In this case, there is no tag
which appears to have a high proportion of missing values when
signal:noise &gt; 10.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_missing_SN_per_sample.html">plot_missing_SN_per_sample</a></span><span class="op">(</span><span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'psms_filtered_norm'</span><span class="op">]</span><span class="op">]</span>, bins <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="TMT_PSM_QC_Summarisation_files/figure-html/unnamed-chunk-9-1.png" alt="Missing values per tag, in relation to the signal:noise ratio" width="75%"><p class="caption">
Missing values per tag, in relation to the signal:noise ratio
</p>
</div>
<p>Based on the above, we will filter the PSMs to only retain those with
S:N &gt; 10 using <code>filter_TMT_PSMs</code> since these PSMs are
clearly much closer to the limit of detection for the MS and therefore
likely to contain less accurate quantification data. Note that this only
necessitates removing ~ 1/20 of the PSMs in this case. Using the same
function, we will also remove PSMs with interference/co-isolation
&gt;50% since these are also likely to contain less accurate
quantification data.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Then filter PSMs to remove low S:N and/or high interference</span></span>
<span><span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'psms_filtered_norm'</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filter_TMT_PSMs.html">filter_TMT_PSMs</a></span><span class="op">(</span><span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'psms_filtered_norm'</span><span class="op">]</span><span class="op">]</span>,</span>
<span>                                                  inter_thresh<span class="op">=</span><span class="fl">50</span>, sn_thresh<span class="op">=</span><span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt; Filtering PSMs...</span></span>
<span><span class="co">#&gt; 4747 features found from 2242 master proteins =&gt; Initial PSMs</span></span>
<span><span class="co">#&gt; 4747 features found from 2242 master proteins =&gt; Removing PSMs without quantification values</span></span>
<span><span class="co">#&gt; 4628 features found from 2206 master proteins =&gt; Removing PSMs with high Co-isolation/interference</span></span>
<span><span class="co">#&gt; 4470 features found from 2178 master proteins =&gt; Removing PSMs with low average S:N ratio</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="summarising-to-protein-level-abundances">Summarising to protein-level abundances<a class="anchor" aria-label="anchor" href="#summarising-to-protein-level-abundances"></a>
</h2>
<p>Now that we have inspected the PSM-level quantification and filtered
the PSMs, we can summarise the PSMs to protein-level abundances.</p>
<p>For PSM to protein summarisation with TMT quantification, simply
summing together the PSM-level abundances provides accurate estimates,
so long as there are no missing values. If many missing values are
present, the <code><a href="https://rdrr.io/pkg/MsCoreUtils/man/robustSummary.html" class="external-link">MsCoreUtils::robustSummary</a></code> method may be
preferred, since it is able to summarise accurately, even in the
presence of missing values <span class="citation">(Sticker et al.
2020)</span>.</p>
<p>We will see both approaches here, since the later approach requires
some additional steps to ensure it is performed appropriately and the
comparison between them is informative when considering what it means to
summarise from PSMs to protein level abundances.</p>
<div class="section level3">
<h3 id="summarisation-with-sum">Summarisation with sum<a class="anchor" aria-label="anchor" href="#summarisation-with-sum"></a>
</h3>
<p>We start by aggregating our PSMs using a simple sum. First, we need
to deal with missing values. If we leave the missing values in, we
either need to allow some protein quantifications to be <code>NA</code>,
or else ignore the missing values in the summarisation, which will
potentially mean different PSMs are summed for the a given protein
across the samples. Here, we will therefore remove all PSMs with missing
values.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'psms_filtered_forSum'</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">QFeatures</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">filterNA</a></span><span class="op">(</span></span>
<span>  <span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'psms_filtered_norm'</span><span class="op">]</span><span class="op">]</span>, <span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<p>Next, we will remove PSMs for proteins with fewer than 2 PSMs. This
is a common filter in proteomics to ensure the protein-level
quantifications are derived from at least two independent observations.
In some cases, for example phosphoproteomics, or where this filter
appears to be too stringent, it may be appropriate to skip it.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">min_psms</span> <span class="op">&lt;-</span> <span class="fl">2</span></span>
<span><span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'psms_filtered_forSum'</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">biomasslmb</span><span class="fu">::</span><span class="fu"><a href="../reference/filter_features_per_protein.html">filter_features_per_protein</a></span><span class="op">(</span></span>
<span>  <span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'psms_filtered_forSum'</span><span class="op">]</span><span class="op">]</span>, min_features <span class="op">=</span> <span class="va">min_psms</span><span class="op">)</span></span></code></pre></div>
<p>Finally, we perform the summarisation.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Aggregate to protein-level abundances (using QFeatures function)</span></span>
<span><span class="va">tmt_qf</span> <span class="op">&lt;-</span> <span class="fu">QFeatures</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">aggregateFeatures</a></span><span class="op">(</span><span class="va">tmt_qf</span>, </span>
<span>                                       i <span class="op">=</span> <span class="st">"psms_filtered_forSum"</span>, </span>
<span>                                       fcol <span class="op">=</span> <span class="st">"Master.Protein.Accessions"</span>,</span>
<span>                                       name <span class="op">=</span> <span class="st">"protein_sum"</span>,</span>
<span>                                       fun <span class="op">=</span> <span class="fu">base</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Your row data contain missing values. Please read the relevant</span></span>
<span><span class="co">#&gt; section(s) in the aggregateFeatures manual page regarding the effects</span></span>
<span><span class="co">#&gt; of missing values on data aggregation.</span></span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'protein_sum'</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">QFeatures</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-processing.html" class="external-link">logTransform</a></span><span class="op">(</span></span>
<span>  <span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'protein_sum'</span><span class="op">]</span><span class="op">]</span>, base<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="summarisation-with-robustsummary">Summarisation with robustSummary<a class="anchor" aria-label="anchor" href="#summarisation-with-robustsummary"></a>
</h3>
<p>In most instances, for TMT proteomics, summarisation with
<code>sum</code> will be sufficient. However, there may be cases where
the need to remove missing values for <code>sum</code> summarisation
removes too many PSMs. In this case, an alternative summarisation
approach using <code>robustSummary</code> <span class="citation">(Sticker et al. 2020)</span> may be appropriate.</p>
<p>With <code>robustSummary</code>, we do not need to remove all PSMs
with missing values since the summarisation algorithm deals with them
appropriately. However, we still don’t want to retain PSMs with too many
missing values, since these will not be very informative in estimating
the protein-level quantification. Here, we will retain PSMs with at most
2/10 missing values</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'psms_filtered_forRobust'</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">QFeatures</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">filterNA</a></span><span class="op">(</span></span>
<span>  <span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'psms_filtered_norm'</span><span class="op">]</span><span class="op">]</span>, <span class="fl">0.2</span><span class="op">)</span></span></code></pre></div>
<p>Next, we again remove PSMs for proteins with fewer than 2 PSMs.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'psms_filtered_forRobust'</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">biomasslmb</span><span class="fu">::</span><span class="fu"><a href="../reference/filter_features_per_protein.html">filter_features_per_protein</a></span><span class="op">(</span></span>
<span>  <span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'psms_filtered_forRobust'</span><span class="op">]</span><span class="op">]</span>, min_features <span class="op">=</span> <span class="va">min_psms</span><span class="op">)</span></span></code></pre></div>
<p>For the <code>robustSummary</code> summarisation, we need our
quantification values to be approximately Gaussian distributed. Hence,
we log transform them.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'psms_filtered_forRobust'</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">QFeatures</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-processing.html" class="external-link">logTransform</a></span><span class="op">(</span></span>
<span>  <span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'psms_filtered_forRobust'</span><span class="op">]</span><span class="op">]</span>, base<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p>Now we can summarise with <code>robustSummary</code></p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Aggregate to protein-level abundances (using QFeatures function)</span></span>
<span><span class="va">tmt_qf</span> <span class="op">&lt;-</span> <span class="fu">QFeatures</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">aggregateFeatures</a></span><span class="op">(</span><span class="va">tmt_qf</span>, </span>
<span>                                       i <span class="op">=</span> <span class="st">"psms_filtered_forRobust"</span>, </span>
<span>                                       fcol <span class="op">=</span> <span class="st">"Master.Protein.Accessions"</span>,</span>
<span>                                       name <span class="op">=</span> <span class="st">"protein_robust"</span>,</span>
<span>                                       fun <span class="op">=</span> <span class="fu">MsCoreUtils</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/MsCoreUtils/man/robustSummary.html" class="external-link">robustSummary</a></span>,</span>
<span>                                       maxit<span class="op">=</span><span class="fl">1000</span><span class="op">)</span></span>
<span><span class="co">#&gt; Your quantitative and row data contain missing values. Please read the</span></span>
<span><span class="co">#&gt; relevant section(s) in the aggregateFeatures manual page regarding the</span></span>
<span><span class="co">#&gt; effects of missing values on data aggregation.</span></span></code></pre></div>
<p>Prior to summaristaion, we removed PSMs from proteins with fewer than
2 PSMs. However, since we left in PSMs with missing values, it’s
possible for some protein-level abundances to be derived from just a
single PSM. We can use the <code>get_protein_no_quant_mask</code> from
<code>biomasslmb</code> to identify where the protein abundances will be
derived from fewer than <code>n</code> features (PSMs). We can then give
this mask to <code>mask_protein_level_quant</code> to replace these
quantification values with NA.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># plot = TRUE means we will also get a plot of the number of proteins quantified in each sample</span></span>
<span><span class="va">protein_retain_mask</span> <span class="op">&lt;-</span> <span class="fu">biomasslmb</span><span class="fu">::</span><span class="fu"><a href="../reference/get_protein_no_quant_mask.html">get_protein_no_quant_mask</a></span><span class="op">(</span></span>
<span>  <span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'psms_filtered_forRobust'</span><span class="op">]</span><span class="op">]</span>, min_features<span class="op">=</span><span class="fl">2</span>, plot<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span> </span></code></pre></div>
<p><img src="TMT_PSM_QC_Summarisation_files/figure-html/unnamed-chunk-19-1.png" width="50%"></p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'protein_robust'</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">biomasslmb</span><span class="fu">::</span><span class="fu"><a href="../reference/mask_protein_level_quant.html">mask_protein_level_quant</a></span><span class="op">(</span></span>
<span>  <span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'protein_robust'</span><span class="op">]</span><span class="op">]</span>, <span class="va">protein_retain_mask</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="comparing-the-summarisation-approaches">Comparing the summarisation approaches<a class="anchor" aria-label="anchor" href="#comparing-the-summarisation-approaches"></a>
</h3>
<p>We can see that performing the robust summarisation allows us to
retain more proteins, since we can still use the PSMs with a low number
of missing values.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'protein_sum'</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 891</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'protein_robust'</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 917</span></span></code></pre></div>
<p>However, with sum summarisation, we have complete quantification data
(no missing values), whereas with <code>robustSummary</code>, we will
proteins with missing values in some samples. In this case, the
protein-level data is 99.68 % complete, so the impact of the missing
values on the downstream analysis will be minimal. We can inspect
missing values using the <code><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-missing-data.html" class="external-link">QFeatures::nNA</a></code> function, or just
<code>is.na</code> on the quantitative data matrix.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">QFeatures</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-missing-data.html" class="external-link">nNA</a></span><span class="op">(</span><span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'protein_sum'</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">$</span><span class="va">nNAcols</span></span>
<span><span class="co">#&gt; DataFrame with 10 rows and 3 columns</span></span>
<span><span class="co">#&gt;             name       nNA       pNA</span></span>
<span><span class="co">#&gt;      &lt;character&gt; &lt;integer&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; 1  Abundance....         0         0</span></span>
<span><span class="co">#&gt; 2  Abundance....         0         0</span></span>
<span><span class="co">#&gt; 3  Abundance....         0         0</span></span>
<span><span class="co">#&gt; 4  Abundance....         0         0</span></span>
<span><span class="co">#&gt; 5  Abundance....         0         0</span></span>
<span><span class="co">#&gt; 6  Abundance....         0         0</span></span>
<span><span class="co">#&gt; 7  Abundance....         0         0</span></span>
<span><span class="co">#&gt; 8  Abundance....         0         0</span></span>
<span><span class="co">#&gt; 9  Abundance....         0         0</span></span>
<span><span class="co">#&gt; 10 Abundance....         0         0</span></span>
<span><span class="fu">QFeatures</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-missing-data.html" class="external-link">nNA</a></span><span class="op">(</span><span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'protein_robust'</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">$</span><span class="va">nNAcols</span></span>
<span><span class="co">#&gt; DataFrame with 10 rows and 3 columns</span></span>
<span><span class="co">#&gt;             name       nNA        pNA</span></span>
<span><span class="co">#&gt;      &lt;character&gt; &lt;integer&gt;  &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; 1  Abundance....         0 0.00000000</span></span>
<span><span class="co">#&gt; 2  Abundance....         0 0.00000000</span></span>
<span><span class="co">#&gt; 3  Abundance....         1 0.00109051</span></span>
<span><span class="co">#&gt; 4  Abundance....         4 0.00436205</span></span>
<span><span class="co">#&gt; 5  Abundance....         1 0.00109051</span></span>
<span><span class="co">#&gt; 6  Abundance....         3 0.00327154</span></span>
<span><span class="co">#&gt; 7  Abundance....         0 0.00000000</span></span>
<span><span class="co">#&gt; 8  Abundance....         2 0.00218103</span></span>
<span><span class="co">#&gt; 9  Abundance....         7 0.00763359</span></span>
<span><span class="co">#&gt; 10 Abundance....        11 0.01199564</span></span>
<span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">message</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">'%.2f %% missing values'</span>, <span class="fl">100</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'protein_robust'</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; 0.32 % missing values</span></span></code></pre></div>
<p>To better understand the difference between how the two summarisation
methods use the available data, we can inspect the single protein which
has no missing values with <code>robustSummary</code>, but is not
summarised with <code>sum</code>.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Identify the proteins with no missing values present only following robustSummary</span></span>
<span><span class="va">protein_robust_proteins</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">filterNA</a></span><span class="op">(</span><span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'protein_robust'</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">protein_sum_proteins</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">filterNA</a></span><span class="op">(</span><span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'protein_sum'</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">no_missing_robustSummary_only</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/setops.html" class="external-link">setdiff</a></span><span class="op">(</span><span class="va">protein_robust_proteins</span>, <span class="va">protein_sum_proteins</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">no_missing_robustSummary_only</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "P22695"</span></span></code></pre></div>
<p>If we inspect the PSM level abundances for this protein, we see that
there are 3 PSMs, of which the first two contain missing values in a
single sample. Thus, we have only one complete PSM, which is
insufficient for confident summarisation using <code>sum</code>.
However, all 3 PSMs have sufficiently few missing values to be used with
the <code>robustSummary</code> method and all samples have at least two
non missing values across the 3 PSMs, so this protein is quantified in
all samples with <code>robustSummary</code>.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">retain_psms</span> <span class="op">&lt;-</span> <span class="fu">rowData</span><span class="op">(</span></span>
<span>  <span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'psms_filtered_norm'</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">$</span><span class="va">Master.Protein.Accessions</span><span class="op">==</span><span class="va">no_missing_robustSummary_only</span></span>
<span></span>
<span><span class="fu">assay</span><span class="op">(</span><span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'psms_filtered_norm'</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="va">retain_psms</span>,<span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;      Abundance.126 Abundance.127N Abundance.127C Abundance.128N Abundance.128C</span></span>
<span><span class="co">#&gt; 517      114.68442       158.2849       286.0687       438.4889       173.5508</span></span>
<span><span class="co">#&gt; 1686     587.11851       412.1598       322.1113       232.8183      1419.0267</span></span>
<span><span class="co">#&gt; 3597      89.15436       134.5185       212.4164       336.5223       139.6542</span></span>
<span><span class="co">#&gt;      Abundance.129N Abundance.129C Abundance.130N Abundance.130C Abundance.131</span></span>
<span><span class="co">#&gt; 517        41.27858       15.23801       10.07546             NA            NA</span></span>
<span><span class="co">#&gt; 1686      494.28845             NA       23.85513      33.861821     58.181084</span></span>
<span><span class="co">#&gt; 3597       31.63687       24.65788       17.03938       4.108757      8.366749</span></span></code></pre></div>
<p>We can directly compare the protein-level abundance estimates to
explore where these two summarisation methods differ in their
estimates.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Define a function to extract the protein abundances in long form and</span></span>
<span><span class="co"># add a column annotating the method</span></span>
<span><span class="va">get_long_form_prot_exp</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">obj</span>, <span class="va">method_name</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu">assay</span><span class="op">(</span><span class="va">obj</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">rownames_to_column</a></span><span class="op">(</span><span class="st">'protein'</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html" class="external-link">pivot_longer</a></span><span class="op">(</span>cols<span class="op">=</span><span class="op">-</span><span class="va">protein</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>method <span class="op">=</span> <span class="va">method_name</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># Single object with protein inference from both methods </span></span>
<span><span class="va">compare_protein_abundances</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span><span class="fu">longFormat</span><span class="op">(</span><span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'protein_sum'</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, method<span class="op">=</span><span class="st">'Sum'</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span><span class="fu">longFormat</span><span class="op">(</span><span class="va">tmt_qf</span><span class="op">[[</span><span class="st">'protein_robust'</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, method<span class="op">=</span><span class="st">'Robust'</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">compare_protein_abundances</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;   rowname       colname     value method</span></span>
<span><span class="co">#&gt; 1  A1L0T0 Abundance.126  8.340996    Sum</span></span>
<span><span class="co">#&gt; 2  A4D1E9 Abundance.126  9.674770    Sum</span></span>
<span><span class="co">#&gt; 3  A6NHR9 Abundance.126 10.319013    Sum</span></span>
<span><span class="co">#&gt; 4  O00116 Abundance.126  6.884527    Sum</span></span>
<span><span class="co">#&gt; 5  O00154 Abundance.126  5.597842    Sum</span></span>
<span><span class="co">#&gt; 6  O00159 Abundance.126 10.410898    Sum</span></span></code></pre></div>
<p>We want to identify the proteins with the lowest correlation for
abundances across the samples between the two methods.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">proteins_of_interest</span> <span class="op">&lt;-</span> <span class="va">compare_protein_abundances</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_wider.html" class="external-link">pivot_wider</a></span><span class="op">(</span>names_from <span class="op">=</span> <span class="va">method</span>, values_from <span class="op">=</span> <span class="va">value</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">rowname</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>cor<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor</a></span><span class="op">(</span><span class="va">Sum</span>, <span class="va">Robust</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice_min</a></span><span class="op">(</span><span class="va">cor</span>, n<span class="op">=</span><span class="fl">5</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">rowname</span><span class="op">)</span></span></code></pre></div>
<p>Below, we define a function to plot the peptide and protein level
abundances for a single protein.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plot_pep_and_protein</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">protein_of_interest</span><span class="op">)</span> <span class="op">{</span></span>
<span>  </span>
<span>  <span class="va">to_plot_compare</span> <span class="op">&lt;-</span> <span class="va">compare_protein_abundances</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">rowname</span> <span class="op">==</span> <span class="va">protein_of_interest</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="fu">QFeatures</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/filterFeatures.html" class="external-link">filterFeatures</a></span><span class="op">(</span></span>
<span>    <span class="va">tmt_qf</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-filtering.html" class="external-link">VariableFilter</a></span><span class="op">(</span><span class="st">"Master.Protein.Accessions"</span>,</span>
<span>                   <span class="va">protein_of_interest</span>,</span>
<span>                   condition <span class="op">=</span> <span class="st">"=="</span><span class="op">)</span><span class="op">)</span><span class="op">[[</span><span class="st">'psms_filtered_norm'</span><span class="op">]</span><span class="op">]</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">longFormat</span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">colname</span>, y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="va">value</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>group <span class="op">=</span> <span class="va">rowname</span><span class="op">)</span>, colour <span class="op">=</span> <span class="st">'grey'</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="st">'grey'</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">to_plot_compare</span>,</span>
<span>              <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">colname</span>, y <span class="op">=</span> <span class="va">value</span>, colour <span class="op">=</span> <span class="va">method</span>, group <span class="op">=</span> <span class="va">method</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">to_plot_compare</span>,</span>
<span>               <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">colname</span>, y <span class="op">=</span> <span class="va">value</span>, colour <span class="op">=</span> <span class="va">method</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_colour_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="../reference/get_cat_palette.html">get_cat_palette</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span>,</span>
<span>                        name <span class="op">=</span> <span class="st">'LFQ summarisation method'</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="../reference/theme_biomasslmb.html">theme_biomasslmb</a></span><span class="op">(</span>base_size <span class="op">=</span> <span class="fl">15</span>, border <span class="op">=</span> <span class="cn">FALSE</span>, base_family <span class="op">=</span> <span class="st">'sans'</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, vjust <span class="op">=</span> <span class="fl">1</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span></span>
<span>      title <span class="op">=</span> <span class="va">protein_of_interest</span>,</span>
<span>      x <span class="op">=</span> <span class="st">''</span>,</span>
<span>      y <span class="op">=</span> <span class="st">'Protein abundance (log2)'</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>In the plots below, PSM level quantifications are in grey and the
summarised protein-level abundances are in colours. You can disregard
the y-axis scale when comparing between the methods, or between the
protein and peptide level abundances. The important thing to focus on is
the abundance profile between the tags for any one summarisation method
or PSM.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">for</span><span class="op">(</span><span class="va">poi</span> <span class="kw">in</span> <span class="va">proteins_of_interest</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu">plot_pep_and_protein</span><span class="op">(</span><span class="va">poi</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="TMT_PSM_QC_Summarisation_files/figure-html/unnamed-chunk-27-1.png" width="45%"><img src="TMT_PSM_QC_Summarisation_files/figure-html/unnamed-chunk-27-2.png" width="45%"><img src="TMT_PSM_QC_Summarisation_files/figure-html/unnamed-chunk-27-3.png" width="45%"><img src="TMT_PSM_QC_Summarisation_files/figure-html/unnamed-chunk-27-4.png" width="45%"><img src="TMT_PSM_QC_Summarisation_files/figure-html/unnamed-chunk-27-5.png" width="45%"></p>
<p>In the first 4 cases, the protein-level quantification values are
relatively similar, with the exception of the final tag, where we see
clear disagreement between the PSMs for how abundant the tag is relative
to the other tags. Since <code>sum</code> simply adds together the
abundance values across the PSMs for a given tag and the abundance
values may be across multiple orders of magnitude, the protein-level
abundance pattern across the tags is weighted towards the most abundant
PSMs. This is likely to be a positive attribute, since the most abundant
PSMs will be more accurately quantified. However, the summarisation is
also sensitive to high abundance outliers. On the other hand,
<code>robustSummary</code> is less sensitive to outliers, but it treats
all PSMs equally. In the final example (P29692) we can see how this can
lead to potential inaccuracies when we have only 2 PSMs and one is very
low intensity. For example, tag 129C is mid-range intensity compared to
the other tags for the abundant PSM, but very low for the less abundant
PSM. This has almost no impact on the <code>sum</code> summarise protein
abundance, but a large impact when we use <code>robustSummary</code></p>
</div>
<div class="section level3">
<h3 id="summary">Summary<a class="anchor" aria-label="anchor" href="#summary"></a>
</h3>
<p>Here, we have QCed the PSMs and filtered them in the following
steps:</p>
<ul>
<li>Remove PSMs that are likely from contaminants</li>
<li>Remove PSMs without a unique ‘master protein’</li>
<li>Remove PSMs without any quantification values</li>
<li>Normalise the abundance values so they have the same median value in
all samples</li>
<li>Remove PSMs with very low signal:noise</li>
<li>Remove PSMs with high interference</li>
</ul>
<p>We then summarised the PSM level abundances with two approaches,
<code>sum</code> and <code>robustSummary</code>, before comparing the
protein-level abundances obtained with each of them.
<code>robustSummary</code> handles missing values and therefore enables
more proteins to be quantified. <code>sum</code> is sensitive to
outliers with very high abundance, while <code>robust</code> is more
sensitive to outliers with very low abundance.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.4.2 (2024-10-31)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 24.04.1 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] dplyr_1.1.4                 tidyr_1.3.1                </span></span>
<span><span class="co">#&gt;  [3] ggplot2_3.5.1               biomasslmb_0.0.1           </span></span>
<span><span class="co">#&gt;  [5] QFeatures_1.16.0            MultiAssayExperiment_1.32.0</span></span>
<span><span class="co">#&gt;  [7] SummarizedExperiment_1.36.0 Biobase_2.66.0             </span></span>
<span><span class="co">#&gt;  [9] GenomicRanges_1.58.0        GenomeInfoDb_1.42.3        </span></span>
<span><span class="co">#&gt; [11] IRanges_2.40.1              S4Vectors_0.44.0           </span></span>
<span><span class="co">#&gt; [13] BiocGenerics_0.52.0         MatrixGenerics_1.18.1      </span></span>
<span><span class="co">#&gt; [15] matrixStats_1.5.0          </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] DBI_1.2.3               rlang_1.1.5             magrittr_2.0.3         </span></span>
<span><span class="co">#&gt;  [4] clue_0.3-66             compiler_4.4.2          RSQLite_2.3.9          </span></span>
<span><span class="co">#&gt;  [7] png_0.1-8               systemfonts_1.2.1       vctrs_0.6.5            </span></span>
<span><span class="co">#&gt; [10] reshape2_1.4.4          stringr_1.5.1           ProtGenerics_1.38.0    </span></span>
<span><span class="co">#&gt; [13] pkgconfig_2.0.3         crayon_1.5.3            fastmap_1.2.0          </span></span>
<span><span class="co">#&gt; [16] XVector_0.46.0          labeling_0.4.3          rmarkdown_2.29         </span></span>
<span><span class="co">#&gt; [19] UCSC.utils_1.2.0        visdat_0.6.0            ragg_1.3.3             </span></span>
<span><span class="co">#&gt; [22] purrr_1.0.4             bit_4.5.0.1             xfun_0.50              </span></span>
<span><span class="co">#&gt; [25] zlibbioc_1.52.0         cachem_1.1.0            jsonlite_1.8.9         </span></span>
<span><span class="co">#&gt; [28] blob_1.2.4              DelayedArray_0.32.0     cluster_2.1.6          </span></span>
<span><span class="co">#&gt; [31] R6_2.6.1                bslib_0.9.0             stringi_1.8.4          </span></span>
<span><span class="co">#&gt; [34] genefilter_1.88.0       jquerylib_0.1.4         Rcpp_1.0.14            </span></span>
<span><span class="co">#&gt; [37] knitr_1.49              BiocBaseUtils_1.8.0     Matrix_1.7-1           </span></span>
<span><span class="co">#&gt; [40] splines_4.4.2           igraph_2.1.4            tidyselect_1.2.1       </span></span>
<span><span class="co">#&gt; [43] abind_1.4-8             yaml_2.3.10             lattice_0.22-6         </span></span>
<span><span class="co">#&gt; [46] tibble_3.2.1            plyr_1.8.9              withr_3.0.2            </span></span>
<span><span class="co">#&gt; [49] KEGGREST_1.46.0         evaluate_1.0.3          desc_1.4.3             </span></span>
<span><span class="co">#&gt; [52] survival_3.7-0          Biostrings_2.74.1       pillar_1.10.1          </span></span>
<span><span class="co">#&gt; [55] corrplot_0.95           generics_0.1.3          munsell_0.5.1          </span></span>
<span><span class="co">#&gt; [58] scales_1.3.0            xtable_1.8-4            glue_1.8.0             </span></span>
<span><span class="co">#&gt; [61] lazyeval_0.2.2          tools_4.4.2             robustbase_0.99-4-1    </span></span>
<span><span class="co">#&gt; [64] annotate_1.84.0         fs_1.6.5                XML_3.99-0.18          </span></span>
<span><span class="co">#&gt; [67] grid_4.4.2              cutr_0.0.0.9000         MsCoreUtils_1.18.0     </span></span>
<span><span class="co">#&gt; [70] AnnotationDbi_1.68.0    colorspace_2.1-1        GenomeInfoDbData_1.2.13</span></span>
<span><span class="co">#&gt; [73] naniar_1.1.0            cli_3.6.4               textshaping_1.0.0      </span></span>
<span><span class="co">#&gt; [76] S4Arrays_1.6.0          AnnotationFilter_1.30.0 gtable_0.3.6           </span></span>
<span><span class="co">#&gt; [79] DEoptimR_1.1-3-1        sass_0.4.9              digest_0.6.37          </span></span>
<span><span class="co">#&gt; [82] SparseArray_1.6.1       htmlwidgets_1.6.4       farver_2.1.2           </span></span>
<span><span class="co">#&gt; [85] memoise_2.0.1           htmltools_0.5.8.1       pkgdown_2.1.1          </span></span>
<span><span class="co">#&gt; [88] lifecycle_1.0.4         httr_1.4.7              bit64_4.6.0-1          </span></span>
<span><span class="co">#&gt; [91] MASS_7.3-61</span></span></code></pre></div>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-http://zotero.org/users/5634351/items/ZK69WYZ2" class="csl-entry">
McAlister, Graeme C., David P. Nusinow, Mark P. Jedrychowski, Martin
Wühr, Edward L. Huttlin, Brian K. Erickson, Ramin Rad, Wilhelm Haas, and
Steven P. Gygi. 2014. <span>“MultiNotch MS3 Enables Accurate, Sensitive,
and Multiplexed Detection of Differential Expression Across Cancer Cell
Line Proteomes.”</span> <em>Analytical Chemistry</em> 86 (14): 7150–58.
<a href="https://doi.org/10.1021/ac502040v" class="external-link">https://doi.org/10.1021/ac502040v</a>.
</div>
<div id="ref-http://zotero.org/users/5634351/items/ME322P2A" class="csl-entry">
O’Connell, Jeremy D., Joao A. Paulo, Jonathon J. O’Brien, and Steven P.
Gygi. 2018. <span>“Proteome-Wide Evaluation of Two Common Protein
Quantification Methods.”</span> <em>Journal of Proteome Research</em> 17
(5): 1934–42. <a href="https://doi.org/10.1021/acs.jproteome.8b00016" class="external-link">https://doi.org/10.1021/acs.jproteome.8b00016</a>.
</div>
<div id="ref-http://zotero.org/users/5634351/items/FZN3QTTZ" class="csl-entry">
Sticker, Adriaan, Ludger Goeminne, Lennart Martens, and Lieven Clement.
2020. <span>“Robust Summarization and Inference in Proteome-wide
Label-free Quantification.”</span> <em>Molecular &amp; cellular
proteomics: MCP</em> 19 (7): 1209–19. <a href="https://doi.org/10.1074/mcp.RA119.001624" class="external-link">https://doi.org/10.1074/mcp.RA119.001624</a>.
</div>
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Tom Smith.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
